<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>Plato - app/scripts/common/common.js</title>

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="description" content="">
  <meta name="author" content="">

  <!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->

  <link href="../../assets/css/vendor/morris.css" rel="stylesheet">
  <link href="../../assets/css/vendor/bootstrap.css" rel="stylesheet">
  <link href="../../assets/css/vendor/font-awesome.css" rel="stylesheet">
  <link href="../../assets/css/vendor/font-awesome-ie7.css" rel="stylesheet">
  <link href="../../assets/css/vendor/codemirror.css" rel="stylesheet">
  <link href="../../assets/css/plato.css" rel="stylesheet">
  <link href="../../assets/css/plato-file.css" rel="stylesheet">

</head>

<body>

<div class="navbar navbar-fixed-top">
  <div class="container">
    <a class="navbar-brand" href="http://github.com/es-analysis/plato">Plato on Github</a>
    <ul class="nav navbar-nav">
      <li>
        <a href="../../index.html">Report Home</a>
      </li>
    </ul>
  </div>
</div>

<div class="jumbotron">
  <div class="container">
    <h1>app/scripts/common/common.js</h1>
  </div>
</div>

<div class="container aggregate-stats">
  <div class="row">
    <div class="col-md-6">
      <h2 class="header">Maintainability <a href="http://blogs.msdn.com/b/codeanalysis/archive/2007/11/20/maintainability-index-range-and-meaning.aspx"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="A value between 0 and 100 that represents the relative ease of maintaining the code. A high value means better maintainability." data-original-title="Maintainability Index"  data-container="body"></i></a></h2>
      <p class="stat">61.23</p>
    </div>
    <div class="col-md-6">
      <h2 class="header">Lines of code <i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Source Lines of Code / Logical Lines of Code" data-original-title="SLOC/LSLOC" data-container="body"></i></h2>
      <p class="stat">476</p>
    </div>
  </div>
  <div class="row historical">
    <div class="col-md-6">
      <p id="chart_historical_maint" class="chart"></p>
    </div>
    <div class="col-md-6">
      <p id="chart_historical_sloc" class="chart"></p>
    </div>
  </div>
  <div class="row">
    <div class="col-md-6">
      <h2 class="header">Difficulty  <a href="http://en.wikipedia.org/wiki/Halstead_complexity_measures"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="The difficulty measure is related to the difficulty of the program to write or understand." data-original-title="Difficulty" data-container="body"></i></a></h2>
      <p class="stat">25.71</p>
    </div>
    <div class="col-md-6">
      <h2 class="header">Estimated Errors  <a href="http://en.wikipedia.org/wiki/Halstead_complexity_measures"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Halstead's delivered bugs is an estimate for the number of errors in the implementation." data-original-title="Delivered Bugs" data-container="body"></i></a></h2>
      <p class="stat">4.25</p>
    </div>
  </div>
</div>

<div class="container charts">
  <div class="row">
    <h2 class="header">Function weight</h2>
  </div>
  <div class="row">
    <div class="col-md-6">
      <h3 class="chart-header">By Complexity <a href="http://en.wikipedia.org/wiki/Cyclomatic_complexity"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="This metric counts the number of distinct paths through a block of code. Lower values are better." data-original-title="Cyclomatic Complexity" data-container="body"></i></a></h3>
      <div id="fn-by-complexity" class="stat"></div>
    </div>
    <div class="col-md-6">
      <h3 class="chart-header">By SLOC  <i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Source Lines of Code / Logical Lines of Code" data-original-title="SLOC/LSLOC" data-container="body"></i></h3>
      <div id="fn-by-sloc" class="stat"></div>
    </div>
  </div>
</div>

<div class="container">
  <div class="row">
    <textarea id="file-source" class="col-md-12">/*
 * Copyright 2013(c) The Ontario Institute for Cancer Research. All rights reserved.
 *
 * This program and the accompanying materials are made available under the terms of the GNU Public
 * License v3.0. You should have received a copy of the GNU General Public License along with this
 * program. If not, see &lt;http://www.gnu.org/licenses/&gt;.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS &quot;AS IS&quot; AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
(function () {
  &#039;use strict&#039;;

  var module = angular.module(&#039;app.common&#039;, [
    &#039;app.common.services&#039;,
    &#039;app.common.footer&#039;,
    &#039;icgc.common.version&#039;,
    &#039;icgc.common.notify&#039;,
    &#039;icgc.common.location&#039;
  ]);

  module.filter(&#039;sum&#039;, function () {
    return function (items, param) {
      var ret = null;
      if (angular.isArray(items)) {
        ret = _.reduce(_.pluck(items, param), function (sum, num) {
          return sum + num;
        });
      }
      return ret;
    };
  });

  module.filter(&#039;startsWith&#039;, function () {
    return function (string, start) {
      var ret = null;
      if (angular.isString(string)) {
        ret = string.indexOf(start) === 0 ? string : null;
      }
      return ret;
    };
  });

  module.filter(&#039;numberPT&#039;, function ($filter) {
    return function (number) {
      if (angular.isNumber(number)) {
        return $filter(&#039;number&#039;)(number);
      } else {
        return number;
      }
    };
  });

  module.filter(&#039;trans&#039;, function () {
    return function (text, capitalize) {
      var translations;

      capitalize = capitalize || false;

      function human(text) {
        var t, words, cWords = [];
        if (!angular.isDefined(text) || (angular.isString(text) &amp;&amp; !text.length)) {
          return &#039;--&#039;;
        }
        if (!angular.isString(text)) {
          return text;
        }

        t = text.replace(/_/g, &#039; &#039;).replace(/^\s+|\s+$/g, &#039;&#039;);
        if (capitalize) {
          words = t.split(&#039; &#039;);
          words.forEach(function (word) {
            cWords.push(word.charAt(0).toUpperCase() + word.slice(1));
          });
          t = cWords.join(&#039; &#039;);
        }
        return t;
      }

      translations = {
        // No Data
        &#039;null&#039;: &#039;No Data&#039;,

        // Facet Titles
        &#039;id&#039;: &#039;Project&#039;,
        &#039;projectId&#039;: &#039;Project&#039;,
        &#039;primarySite&#039;: &#039;Primary Site&#039;,
        &#039;primaryCountries&#039;: &#039;Countries&#039;,
//      &#039;donor_sex&#039;: &#039;Gender&#039;,
        &#039;tumourStageAtDiagnosis&#039;: &#039;Tumour Stage&#039;,
        &#039;vitalStatus&#039;: &#039;Vital Status&#039;,
        &#039;diseaseStatusLastFollowup&#039;: &#039;Disease Status&#039;,
        &#039;relapseType&#039;: &#039;Release Type&#039;,
        &#039;ageAtDiagnosisGroup&#039;: &#039;Age at Diagnosis&#039;,
        &#039;availableDataTypes&#039;: &#039;Available Data Types&#039;,
        &#039;analysisTypes&#039;: &#039;Analysis Types&#039;,
//      &#039;_gene_id&#039;: &#039;Gene ID&#039;,
//      &#039;biotype&#039;: &#039;Type&#039;,
//      &#039;gene_location&#039;: &#039;Location&#039;,

        // Data Types
        &#039;ssm&#039;: &#039;SSM&#039;,
        &#039;cnsm&#039;: &#039;CNSM&#039;,
        &#039;stsm&#039;: &#039;StSM&#039;,
        &#039;sgv&#039;: &#039;SGV&#039;,
        &#039;cngv&#039;: &#039;CNGV&#039;,
        &#039;stgv&#039;: &#039;StGV&#039;,
        &#039;meth&#039;: &#039;METH&#039;,
        &#039;exp&#039;: &#039;EXP&#039;,
        &#039;pexp&#039;: &#039;PEXP&#039;,
        &#039;mirna&#039;: &#039;miRNA&#039;,
        &#039;jcn&#039;: &#039;JCN&#039;,

        // biotype
        &#039;lincRNA&#039;: &#039;lincRNA&#039;,
        &#039;miRNA&#039;: &#039;miRNA&#039;,
        &#039;snRNA&#039;: &#039;snRNA&#039;,
        &#039;snoRNA&#039;: &#039;snoRNA&#039;,
        &#039;rRNA&#039;: &#039;rRNA&#039;,
        &#039;3prime_overlapping_ncrna&#039;: &#039;3\&#039; Overlapping ncRNA&#039;,
        &#039;Mt_rRNA&#039;: &#039;Mt rRNA&#039;,

        // SSM mutation types
        &#039;mutation_type&#039;: &#039;Type&#039;,
        &#039;single base substitution&#039;: &#039;Substitution&#039;,
        &#039;insertion of &lt;=200bp&#039;: &#039;Insertion&#039;,
        &#039;deletion of &lt;=200bp&#039;: &#039;Deletion&#039;,
        &#039;multiple base substitution (&gt;=2bp and &lt;=200bp)&#039;: &#039;MSub&#039;,

        // SSM mutation consequences
        &#039;consequence_type&#039;: &#039;Consequence&#039;,
        &#039;frameshift_variant&#039;: &#039;Frameshift&#039;,
        &#039;missense&#039;: &#039;Missense&#039;,
        &#039;non_conservative_missense_variant&#039;: &#039;Non Conservative Missense&#039;,
        &#039;initiator_codon_variant&#039;: &#039;Initiator Codon&#039;,
        &#039;stop_gained&#039;: &#039;Stop Gained&#039;,
        &#039;stop_lost&#039;: &#039;Stop Lost&#039;,
        &#039;start_gained&#039;: &#039;Start Gained&#039;,
        &#039;exon_lost&#039;: &#039;Exon Lost&#039;,
        &#039;coding_sequence_variant&#039;: &#039;Coding Seq&#039;,
        &#039;inframe_deletion&#039;: &#039;Inframe Del&#039;,
        &#039;inframe_insertion&#039;: &#039;Inframe Ins&#039;,
        &#039;splice_region_variant&#039;: &#039;Splice&#039;,
        &#039;regulatory_region_variant&#039;: &#039;Regulatory&#039;,
        &#039;micro_rna&#039;: &#039;micro RNA&#039;,
        &#039;non_coding_exon_variant&#039;: &#039;NC Exon&#039;,
        &#039;nc_transcript_variant&#039;: &#039;NC Transcript&#039;,
        &#039;5_prime_UTR_variant&#039;: &#039;5\&#039; UTR&#039;,
//      &#039;five_prime_UTR&#039;: &#039;5\&#039; UTR&#039;,
        &#039;upstream_gene_variant&#039;: &#039;Upstream&#039;,
        &#039;synonymous_variant&#039;: &#039;Syn&#039;,
        &#039;stop_retained_variant&#039;: &#039;Stop Retained&#039;,
        &#039;3_prime_UTR_variant&#039;: &#039;3\&#039; UTR&#039;,
        &#039;downstream_gene_variant&#039;: &#039;Downstream&#039;,
        &#039;intron_variant&#039;: &#039;Intron&#039;,
        &#039;intergenic_region&#039;: &#039;Intergenic&#039;,
//      &#039;intergenic_variant&#039;: &#039;intergenic&#039;,
        &#039;intragenic_variant&#039;: &#039;Intragenic&#039;
      };

      return translations.hasOwnProperty(text) ? translations[text] : human(text);
    };
  });

  module.filter(&#039;define&#039;, function () {
    return function (text, nullable) {
      var definitions;

      nullable = nullable || false;

      definitions = {
        // Projects
        &#039;ALL-US&#039;: &#039;Acute Lymphoblastic Leukemia - US&#039;,
        &#039;BLCA-US&#039;: &#039;Bladder Urothelial Cancer - TGCA, US&#039;,
        &#039;BOCA-UK&#039;: &#039;Bone Cancer - UK&#039;,
        &#039;BRCA-EU&#039;: &#039;Breast ER+ and HER2- Cancer - EU/UK&#039;,
        &#039;BRCA-FR&#039;: &#039;Breast Cancer - FR&#039;,
        &#039;BRCA-KR&#039;: &#039;Breast Cancer - KR&#039;,
        &#039;BRCA-MX&#039;: &#039;Breast Cancer - SIGMA, MX&#039;,
        &#039;BRCA-UK&#039;: &#039;Breast Triple Negative/Lobular Cancer - UK&#039;,
        &#039;BRCA-US&#039;: &#039;Breast Cancer - TCGA, US&#039;,
        &#039;CESC-US&#039;: &#039;Cervical Squamous Cell Carcinoma - TCGA, US&#039;,
        &#039;CLLE-ES&#039;: &#039;Chronic Lymphocyclic Leukemia - ES&#039;,
        &#039;CMDI-UK&#039;: &#039;Chronic Myeloid Disorders - UK&#039;,
        &#039;COAD-US&#039;: &#039;Colon Adenocarcinoma - TCGA, US&#039;,
        &#039;COCA-CN&#039;: &#039;Colorectal Cancer - CN&#039;,
        &#039;EOPC-DE&#039;: &#039;Early Onset Prostate Cancer - DE&#039;,
        &#039;ESAD-UK&#039;: &#039;Esophageal Adenocarcinoma - UK&#039;,
        &#039;ESCA-CN&#039;: &#039;Esophageal Cancer - CN&#039;,
        &#039;GACA-CN&#039;: &#039;Gastric Cancer - CN&#039;,
        &#039;GBM-US&#039;: &#039;Brain Glioblastoma Multiforme - TCGA, US&#039;,
        &#039;HNCA-MX&#039;: &#039;Head and Neck Cancer - SIGMA, MX&#039;,
        &#039;HNSC-US&#039;: &#039;Head and Neck Squamous Cell Carcinoma - TCGA, US&#039;,
        &#039;KIRC-US&#039;: &#039;Kidney Renal Clear Cell Carcinoma - TCGA, US&#039;,
        &#039;KIRP-US&#039;: &#039;Kidney Renal Papillary Cell Carcinoma - TCGA, US&#039;,
        &#039;LAML-KR&#039;: &#039;Acute Myeloid Leukemia - KR&#039;,
        &#039;LAML-US&#039;: &#039;Acute Myeloid Leukemia - TCGA, US&#039;,
        &#039;LGG-US&#039;: &#039;Brain Lower Grade Gliona - TCGA, US&#039;,
        &#039;LICA-CN&#039;: &#039;Liver Cancer - CN&#039;,
        &#039;LICA-FR&#039;: &#039;Liver Cancer - FR&#039;,
        &#039;LIHC-US&#039;: &#039;Liver Hepatocellular carcinoma - TCGA, US&#039;,
        &#039;LINC-JP&#039;: &#039;Liver Cancer - NCC, JP&#039;,
        &#039;LIRI-JP&#039;: &#039;Liver Cancer - RIKEN, JP&#039;,
        &#039;LUAD-US&#039;: &#039;Lung Adenocarcinoma - TCGA, US&#039;,
        &#039;LUCA-DE&#039;: &#039;Lung Cancer - DE&#039;,
        &#039;LUSC-US&#039;: &#039;Lung Squamous Cell Carcinoma - TCGA, US&#039;,
        &#039;MALY-DE&#039;: &#039;Malignant Lymphoma - DE&#039;,
        &#039;NACA-CN&#039;: &#039;Nasopharyngeal cancer - CN&#039;,
        &#039;NBL-US&#039;: &#039;Brain Neuroblastoma - US&#039;,
        &#039;NHLY-MX&#039;: &#039;Non Hodgkin Lymphoma - SIGMA, MX&#039;,
        &#039;ORCA-IN&#039;: &#039;Oral Cancer - IN&#039;,
        &#039;OV-AU&#039;: &#039;Ovarian Cancer - AU&#039;,
        &#039;OV-US&#039;: &#039;Ovarian Serous Cystadenocarcinoma - TCGA, US&#039;,
        &#039;PAAD-US&#039;: &#039;Pancreatic Cancer - TCGA, US&#039;,
        &#039;PACA-AU&#039;: &#039;Pancreatic Cancer - AU&#039;,
        &#039;PACA-CA&#039;: &#039;Pancreatic Cancer - CA&#039;,
        &#039;PACA-IT&#039;: &#039;Pancreatic Cancer - IT&#039;,
        &#039;PAEN-AU&#039;: &#039;Pancreatic Cancer Endocrine neoplasms - AU&#039;,
        &#039;PBCA-DE&#039;: &#039;Pediatric Brain Cancer - DE&#039;,
        &#039;PEME-CA&#039;: &#039;Pediatric Medulloblastoma - CA&#039;,
        &#039;PRAD-CA&#039;: &#039;Prostate Adenocarcinoma - CA&#039;,
        &#039;PRAD-UK&#039;: &#039;Prostate Adenocarcinoma - UK&#039;,
        &#039;PRAD-US&#039;: &#039;Prostate Adenocarcinoma - TCGA, US&#039;,
        &#039;PRCA-FR&#039;: &#039;Prostate Cancer - FR&#039;,
        &#039;READ-US&#039;: &#039;Rectum Adenocarcinoma - TCGA, US&#039;,
        &#039;RECA-CN&#039;: &#039;Renal cancer - CN&#039;,
        &#039;RECA-EU&#039;: &#039;Renal Cell Cancer - EU/FR&#039;,
        &#039;SKCA-BR&#039;: &#039;Skin Adenocarcinoma - BR&#039;,
        &#039;SKCM-US&#039;: &#039;Skin Cutaneous melanoma - TCGA, US&#039;,
        &#039;STAD-US&#039;: &#039;Gastric Adenocarcinoma - TCGA, US&#039;,
        &#039;THCA-SA&#039;: &#039;Thyroid Cancer - SA&#039;,
        &#039;THCA-US&#039;: &#039;Head and Neck Thyroid Carcinoma - TCGA, US&#039;,
        &#039;UCEC-US&#039;: &#039;Uterine Corpus Endometrial Carcinoma- TCGA, US&#039;,

        // Data Types
        &#039;Clinical&#039;: &#039;Clinical Data&#039;,
        &#039;SSM&#039;: &#039;Simple Somatic Mutation&#039;,
        &#039;CNSM&#039;: &#039;Copy Number Somatic Mutation&#039;,
        &#039;StSM&#039;: &#039;Structural Somatic Mutation&#039;,
        &#039;SGV&#039;: &#039;Simple Germline Variation&#039;,
        &#039;CNGV&#039;: &#039;Copy Number Germline Variation&#039;,
        &#039;StGV&#039;: &#039;Structural Germline Variation&#039;,
        &#039;METH&#039;: &#039;DNA Methylation&#039;,
        &#039;EXP&#039;: &#039;Gene Expression&#039;,
        &#039;PEXP&#039;: &#039;Protein Expression&#039;,
        &#039;miRNA&#039;: &#039;microRNA Expression&#039;,
        &#039;JCN&#039;: &#039;Exon Junction&#039;,

        // SSM mutation types
        &#039;Substitution&#039;: &#039;single base substitution&#039;,
        &#039;Insertion&#039;: &#039;insertion of &lt;=200bp&#039;,
        &#039;Deletion&#039;: &#039;deletion of &lt;=200bp&#039;,
        &#039;MSub&#039;: &#039;multiple base substitution (&gt;=2bp and &lt;=200bp)&#039;,

        // SSM mutation consequences
        &#039;Consequence&#039;: &#039;SO term:consequence_type&#039;,
        &#039;Frameshift&#039;: &#039;SO term: frameshift_variant&#039;,
        &#039;Missense&#039;: &#039;SO term: missense&#039;,
        &#039;Non Conservative Missense&#039;: &#039;SO term: non_conservative_missense_variant&#039;,
        &#039;Initiator Codon&#039;: &#039;SO term: initiator_codon_variant&#039;,
        &#039;Stop Gained&#039;: &#039;SO term: stop_gained&#039;,
        &#039;Stop Lost&#039;: &#039;SO term: stop_lost&#039;,
        &#039;Start Gained&#039;: &#039;SO term: start_gained&#039;,
        &#039;Exon Lost&#039;: &#039;SO term: exon_lost&#039;,
        &#039;Coding Seq&#039;: &#039;SO term: coding_sequence_variant&#039;,
        &#039;Inframe Del&#039;: &#039;SO term: inframe_deletion&#039;,
        &#039;Inframe Ins&#039;: &#039;SO term: inframe_insertion&#039;,
        &#039;Splice&#039;: &#039;SO term: splice_region_variant&#039;,
        &#039;Regulatory&#039;: &#039;SO term: regulatory_region_variant&#039;,
        &#039;micro RNA&#039;: &#039;SO term: micro_rna&#039;,
        &#039;NC Exon&#039;: &#039;SO term: non_coding_exon_variant&#039;,
        &#039;NC Transcript&#039;: &#039;SO term: nc_transcript_variant&#039;,
        &#039;5\&#039; UTR&#039;: &#039;SO term: 5_prime_UTR_variant&#039;,
        &#039;Upstream&#039;: &#039;SO term: upstream_gene_variant&#039;,
        &#039;Syn&#039;: &#039;SO term: synonymous_variant&#039;,
        &#039;Stop Retained&#039;: &#039;SO term: stop_retained_variant&#039;,
        &#039;3\&#039; UTR&#039;: &#039;SO term: 3_prime_UTR_variant&#039;,
        &#039;Downstream&#039;: &#039;SO term: downstream_gene_variant&#039;,
        &#039;Intron&#039;: &#039;SO term: intron_variant&#039;,
        &#039;intergenic&#039;: &#039;SO term: intergenic_variant&#039;,
        &#039;Intragenic&#039;: &#039;SO term: intragenic_variant&#039;,

        // Sequencing analysis types (Sequencing strategy)
        &#039;WGS&#039;: &#039;Whole Genome Sequencing - random sequencing of the whole genome.&#039;,
        &#039;WGA&#039;: &#039;Whole Genome Amplification followed by random sequencing.&#039;,
        &#039;WXS&#039;: &#039;Random sequencing of exonic regions selected from the genome.&#039;,
        &#039;RNA-Seq&#039;: &#039;Random sequencing of whole transcriptome, also known as Whole Transcriptome Shotgun Sequencing, &#039; +
            &#039;or WTSS&#039;,
        &#039;miRNA-Seq&#039;: &#039;Micro RNA sequencing strategy designed to capture post-transcriptional RNA elements and include &#039; +
            &#039;non-coding functional elements.&#039;,
        &#039;ncRNA-Seq&#039;: &#039;Capture of other non-coding RNA types, including post-translation modification types such as &#039; +
            &#039;snRNA (small nuclear RNA) or snoRNA (small nucleolar RNA), or expression regulation types such as siRNA &#039; +
            &#039;(small interfering RNA) or piRNA/piwi/RNA (piwi-interacting RNA).&#039;,
        &#039;WCS&#039;: &#039;Random sequencing of a whole chromosome or other replicon isolated from a genome.&#039;,
        &#039;CLONE&#039;: &#039;Genomic clone based (hierarchical) sequencing.&#039;,
        &#039;POOLCLONE&#039;: &#039;Shotgun of pooled clones (usually BACs and Fosmids).&#039;,
        &#039;AMPLICON&#039;: &#039;Sequencing of overlapping or distinct PCR or RT-PCR products. For example, metagenomic community &#039; +
            &#039;profiling using SSU rRNA.&#039;,
        &#039;CLONEEND&#039;: &#039;Clone end (5\&#039;, 3\&#039;, or both) sequencing.&#039;,
        &#039;FINISHING&#039;: &#039;Sequencing intended to finish (close) gaps in existing coverage.&#039;,
        &#039;ChIP-Seq&#039;: &#039;chromatin immunoprecipitation.&#039;,
        &#039;MNase-Seq&#039;: &#039;following MNase digestion.&#039;,
        &#039;DNase-Hypersensitivity&#039;: &#039;Sequencing of hypersensitive sites, or segments of open chromatin that are more &#039; +
            &#039;readily cleaved by DNaseI.&#039;,
        &#039;Bisulfite-Seq&#039;: &#039;MethylC-seq. Sequencing following treatment of DNA with bisulfite to convert cytosine &#039; +
            &#039;residues to uracil depending on methylation status.&#039;,
        &#039;EST&#039;: &#039;Single pass sequencing of cDNA templates&#039;,
        &#039;FL-cDNA&#039;: &#039;Full-length sequencing of cDNA templates&#039;,
        &#039;CTS&#039;: &#039;Concatenated Tag Sequencing&#039;,
        &#039;MRE-Seq&#039;: &#039;Methylation-Sensitive Restriction Enzyme Sequencing.&#039;,
        &#039;MeDIP-Seq&#039;: &#039;Methylated DNA Immunoprecipitation Sequencing.&#039;,
        &#039;MBD-Seq&#039;: &#039;Methyl CpG Binding Domain Sequencing.&#039;,
        &#039;Tn-Seq&#039;: &#039;Quantitatively determine fitness of bacterial genes based on how many times a purposely seeded &#039; +
            &#039;transposon gets inserted into each gene of a colony after some time.&#039;,
        &#039;VALIDATION&#039;: &#039;CGHub special request: Independent experiment to re-evaluate putative variants.&#039;,
        &#039;FAIRE-seq&#039;: &#039;Formaldehyde Assisted Isolation of Regulatory Elements&#039;,
        &#039;SELEX&#039;: &#039;Systematic Evolution of Ligands by EXponential enrichment&#039;,
        &#039;RIP-Seq&#039;: &#039;Direct sequencing of RNA immunoprecipitates (includes CLIP-Seq, HITS-CLIP and PAR-CLIP).&#039;,
        &#039;ChIA-PET&#039;: &#039;Direct sequencing of proximity-ligated chromatin immunoprecipitates.&#039;,
        &#039;OTHER&#039;: &#039;Library strategy not listed.&#039;
      };

      // If translation found return that
      // else if nullable return null
      // else pass through original text
      return definitions.hasOwnProperty(text) ? definitions[text] : nullable ? null : text;
    };
  });

  module.filter(&#039;highlight&#039;, function () {
    return function (text, search, hide) {
      text = text || &#039;&#039;;
      hide = hide || false;
      if (search) {
        text = angular.isArray(text) ? text.join(&#039;, &#039;) : text.toString();
        // Shrink extra spaces, removes non alpha-numeric chars
        search = search.toString().replace(/\s+/g, &#039; &#039;).replace(/[^a-zA-Z0-9\s-]/g, &#039;&#039;).split(&#039; &#039;);
        for (var i = 0; i &lt; search.length; ++i) {
          text = text.replace(new RegExp(search[i], &#039;gi&#039;), &#039;^$&amp;$&#039;);
        }

        // if match return text
        if (text.indexOf(&#039;^&#039;) !== -1) {
          return text.replace(/\^/g, &#039;&lt;span class=&quot;match&quot;&gt;&#039;).replace(/\$/g, &#039;&lt;/span&gt;&#039;);
        } else { // no match
          if (hide) {
            return &#039;&#039;;
          } // hide
        }
      }

      // return base text if no match and not hiding
      return text;
    };
  });

  module.factory(&#039;debounce&#039;, function ($timeout, $q) {
    return function (func, wait, immediate) {
      var timeout, deferred;
      deferred = $q.defer();

      return function () {
        var context, later, callNow, args;

        context = this;
        args = arguments;
        later = function () {
          timeout = null;
          if (!immediate) {
            deferred.resolve(func.apply(context, args));
            deferred = $q.defer();
          }
        };
        callNow = immediate &amp;&amp; !timeout;
        if (timeout) {
          $timeout.cancel(timeout);
        }
        timeout = $timeout(later, wait);
        if (callNow) {
          deferred.resolve(func.apply(context, args));
          deferred = $q.defer();
        }
        return deferred.promise;
      };
    };
  });

  module.filter(&#039;typecv&#039;, function () {
    return function (type) {
      var types = {
        &#039;gene-centric&#039;: &#039;gene&#039;,
        &#039;mutation-centric&#039;: &#039;mutation&#039;,
        &#039;donor-centric&#039;: &#039;donor&#039;
      };
      return types[type];
    };
  });

  module.filter(&#039;unique&#039;, function () {
    return function (items) {
      var i, set, item;

      set = [];
      if (items &amp;&amp; items.length) {
        for (i = 0; i &lt; items.length; ++i) {
          item = items[i];
          if (set.indexOf(item) === -1) {
            set.push(item);
          }
        }
      }
      return set;
    };
  });

// Convert a non-array item into an array
  module.filter(&#039;makeArray&#039;, function () {
    return function (items) {
      if (angular.isArray(items)) {
        return items;
      }
      return [items];
    };
  });


// Join parallel arrays into a single array
// eg. [ [&#039;a&#039;, &#039;b&#039;, &#039;c&#039;], [&#039;d&#039;, &#039;e&#039;, &#039;f&#039;] ] | joinFields:&#039;&#039;  =&gt; [&#039;ad&#039;, &#039;be&#039;, &#039;cf&#039;]
  module.filter(&#039;joinFields&#039;, function () {
    return function (items, delim, checkEmpty) {
      var i, j, list, tempList, joinedItem, hasEmpty;

      list = [];

      // Normalize
      for (i = 0; i &lt; items.length; i++) {
        if (!angular.isArray(items[i])) {
          items[i] = [items[i]];
        }
      }

      // Join
      for (i = 0; i &lt; items[0].length; i++) {
        tempList = [];
        hasEmpty = false;
        for (j = 0; j &lt; items.length; j++) {
          tempList.push(items[j][i]);
          if (items[j][i] === &#039;&#039;) {
            hasEmpty = true;
          }
        }

        // Skip join if one of the item is empty string
        if (checkEmpty &amp;&amp; checkEmpty === true &amp;&amp; hasEmpty === true) {
          continue;
        }

        joinedItem = tempList.join(delim).trim();
        if (joinedItem &amp;&amp; joinedItem !== &#039;&#039;) {
          list.push(joinedItem);
        }
      }
      return list;
    };
  });

  module.filter(&#039;bytes&#039;, function () {
    return function (input) {
      var sizes = [&#039;B&#039;, &#039;KB&#039;, &#039;MB&#039;, &#039;GB&#039;, &#039;TB&#039;],
          postTxt = 0,
          bytes = input,
          precision = 2;

      if (bytes &lt;= 1024) {
        precision = 0;
      }

      while (bytes &gt;= 1024) {
        postTxt++;
        bytes = bytes / 1024;
      }

      return Number(bytes).toFixed(precision) + &#039; &#039; + sizes[postTxt];
    };
  });
})();</textarea>
  </div>
</div>

<footer class="footer">
  <div class="container">
    <p>.</p>
  </div>
</footer>

<script type="text/html" id="complexity-popover-template">
  <div class="complexity-notice">
    Complexity : {{ complexity.cyclomatic }} <br>
    Length : {{ complexity.halstead.length }} <br>
    Difficulty : {{ complexity.halstead.difficulty.toFixed(2) }} <br>
    Est # bugs : {{ complexity.halstead.bugs.toFixed(2) }}<br>
  </div>
</script>

<script type="text/javascript" src="../../assets/scripts/bundles/core-bundle.js"></script>
<script type="text/javascript" src="../../assets/scripts/bundles/codemirror.js"></script>
<script type="text/javascript" src="../../assets/scripts/codemirror.markpopovertext.js"></script>
<script type="text/javascript" src="report.js"></script>
<script type="text/javascript" src="report.history.js"></script>
<script type="text/javascript" src="../../assets/scripts/plato-file.js"></script>
</body>
</html>
