<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>Plato - app/scripts/modules/genomemaps/viewer.js</title>

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="description" content="">
  <meta name="author" content="">

  <!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->

  <link href="../../assets/css/vendor/morris.css" rel="stylesheet">
  <link href="../../assets/css/vendor/bootstrap.css" rel="stylesheet">
  <link href="../../assets/css/vendor/font-awesome.css" rel="stylesheet">
  <link href="../../assets/css/vendor/font-awesome-ie7.css" rel="stylesheet">
  <link href="../../assets/css/vendor/codemirror.css" rel="stylesheet">
  <link href="../../assets/css/plato.css" rel="stylesheet">
  <link href="../../assets/css/plato-file.css" rel="stylesheet">

</head>

<body>

<div class="navbar navbar-fixed-top">
  <div class="container">
    <a class="navbar-brand" href="http://github.com/es-analysis/plato">Plato on Github</a>
    <ul class="nav navbar-nav">
      <li>
        <a href="../../index.html">Report Home</a>
      </li>
    </ul>
  </div>
</div>

<div class="jumbotron">
  <div class="container">
    <h1>app/scripts/modules/genomemaps/viewer.js</h1>
  </div>
</div>

<div class="container aggregate-stats">
  <div class="row">
    <div class="col-md-6">
      <h2 class="header">Maintainability <a href="http://blogs.msdn.com/b/codeanalysis/archive/2007/11/20/maintainability-index-range-and-meaning.aspx"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="A value between 0 and 100 that represents the relative ease of maintaining the code. A high value means better maintainability." data-original-title="Maintainability Index"  data-container="body"></i></a></h2>
      <p class="stat">63.22</p>
    </div>
    <div class="col-md-6">
      <h2 class="header">Lines of code <i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Source Lines of Code / Logical Lines of Code" data-original-title="SLOC/LSLOC" data-container="body"></i></h2>
      <p class="stat">584</p>
    </div>
  </div>
  <div class="row historical">
    <div class="col-md-6">
      <p id="chart_historical_maint" class="chart"></p>
    </div>
    <div class="col-md-6">
      <p id="chart_historical_sloc" class="chart"></p>
    </div>
  </div>
  <div class="row">
    <div class="col-md-6">
      <h2 class="header">Difficulty  <a href="http://en.wikipedia.org/wiki/Halstead_complexity_measures"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="The difficulty measure is related to the difficulty of the program to write or understand." data-original-title="Difficulty" data-container="body"></i></a></h2>
      <p class="stat">64.10</p>
    </div>
    <div class="col-md-6">
      <h2 class="header">Estimated Errors  <a href="http://en.wikipedia.org/wiki/Halstead_complexity_measures"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Halstead's delivered bugs is an estimate for the number of errors in the implementation." data-original-title="Delivered Bugs" data-container="body"></i></a></h2>
      <p class="stat">5.71</p>
    </div>
  </div>
</div>

<div class="container charts">
  <div class="row">
    <h2 class="header">Function weight</h2>
  </div>
  <div class="row">
    <div class="col-md-6">
      <h3 class="chart-header">By Complexity <a href="http://en.wikipedia.org/wiki/Cyclomatic_complexity"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="This metric counts the number of distinct paths through a block of code. Lower values are better." data-original-title="Cyclomatic Complexity" data-container="body"></i></a></h3>
      <div id="fn-by-complexity" class="stat"></div>
    </div>
    <div class="col-md-6">
      <h3 class="chart-header">By SLOC  <i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Source Lines of Code / Logical Lines of Code" data-original-title="SLOC/LSLOC" data-container="body"></i></h3>
      <div id="fn-by-sloc" class="stat"></div>
    </div>
  </div>
</div>

<div class="container">
  <div class="row">
    <textarea id="file-source" class="col-md-12">/*
 * Copyright 2013(c) The Ontario Institute for Cancer Research. All rights reserved.
 *
 * This program and the accompanying materials are made available under the terms of the GNU Public
 * License v3.0. You should have received a copy of the GNU General Public License along with this
 * program. If not, see &lt;http://www.gnu.org/licenses/&gt;.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS &quot;AS IS&quot; AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

&#039;use strict&#039;;

angular.module(&#039;icgc.modules.genomeviewer&#039;, [&#039;icgc.modules.genomeviewer.header&#039;, &#039;icgc.modules.genomeviewer.service&#039;]);

angular.module(&#039;icgc.modules.genomeviewer&#039;).controller(&#039;GenomeViewerController&#039;, function() {});

angular.module(&#039;icgc.modules.genomeviewer&#039;).directive(&#039;genomeViewer&#039;, function(GMService, $location) {
  return {
    restrict: &#039;A&#039;,
    template: &#039;&lt;div id=&quot;genome-viewer&quot;&gt;&lt;/div&gt;&#039;,
    replace: true,
    controller: &#039;GenomeViewerController&#039;,
    link: function(scope) {
      var genomeViewer, navigationBar, regionObj = new Region({chromosome: 1, start: 1, end: 1}), tracks = {},
        done = false;

      function setup() {
        genomeViewer = genomeViewer || new GenomeViewer({
          border: &#039;none&#039;,
          region: regionObj,
          defaultRegion: regionObj,
          sidePanel: false,
          targetId: &#039;genome-viewer&#039;,
          autoRender: true,
          drawNavigationBar: false,
          drawKaryotypePanel: true,
          drawChromosomePanel: true,
          drawRegionOverviewPanel: true,
          karyotypePanelConfig: {
              collapsed: false,
              collapsible: false
          },
          chromosomePanelConfig: {
              collapsed: false,
              collapsible: false
          },
          version: &#039;&lt;small class=&quot;pull-left&quot;&gt;Powered by &#039; +
            &#039;&lt;a target=&quot;_blank&quot; href=&quot;http://www.genomemaps.org/&quot;&gt;Genome Maps&lt;/a&gt;&lt;/small&gt;&#039;
        });

        navigationBar = new IcgcNavigationBar({
          handlers: {
            &#039;region:change&#039;: function(event) {
              Utils.setMinRegion(event.region, genomeViewer.getSVGCanvasWidth());
              genomeViewer.trigger(&#039;region:change&#039;, event);
            },
            &#039;restoreDefaultRegion:click&#039;: function(event) {
              Utils.setMinRegion(genomeViewer.defaultRegion, genomeViewer.getSVGCanvasWidth());
              event.region = genomeViewer.defaultRegion;
              genomeViewer.trigger(&#039;region:change&#039;, event);
            }
          }
        });
        genomeViewer.on(&#039;region:change&#039;, function(event) {
          if (event.sender !== navigationBar) {
            navigationBar.setRegion(event.region);
          }
        });
        genomeViewer.on(&#039;region:move&#039;, function(event) {
          if (event.sender !== navigationBar) {
            navigationBar.moveRegion(event.region);
          }
        });

        genomeViewer.setNavigationBar(navigationBar);

        genomeViewer.draw();

        // Adding tracks
        tracks.sequence = new SequenceTrack({
          targetId: null,
          id: 1,
          title: &#039;Sequence&#039;,
          histogramZoom: 20,
          transcriptZoom: 50,
          height: 30,
          visibleRange: {start: 99, end: 100},
          featureTypes: FEATURE_TYPES,

          renderer: new SequenceRenderer(),

          dataAdapter: new SequenceAdapter({
            category: &#039;genomic&#039;,
            subCategory: &#039;region&#039;,
            resource: &#039;sequence&#039;,
            species: genomeViewer.species,
            featureCache: {
              gzip: true,
              chunkSize: 1000
            }
          })
        });

        genomeViewer.addTrack(tracks.sequence);

        tracks.icgcGeneTrack = new IcgcGeneTrack({
          targetId: null,
          id: 2,
          title: &#039;Genes&#039;,
          histogramZoom: 20,
          transcriptZoom: 40,
          height: 260,
          visibleRange: {start: 0, end: 100},
          featureTypes: FEATURE_TYPES,

          renderer: new GeneRenderer({
            handlers: {
              &#039;feature:click&#039;: function (e) {
                var path = &#039;/genes/&#039; + e.feature[e.featureType === &#039;gene&#039; ? &#039;id&#039; : &#039;geneId&#039;];
                scope.$apply(function () {
                  $location.path(path).search({});
                });
              }
            }
          }),

          dataAdapter: new IcgcGeneAdapter({
            category: &#039;genomic&#039;,
            subCategory: &#039;region&#039;,
            resource: &#039;gene&#039;,
            species: genomeViewer.species,
            featureCache: {
              gzip: true,
              chunkSize: 10000
            },
            filters: {},
            options: {
              histogramThreshold: 20 * 1000 * 1000,
              transcriptThreshold: 10 * 1000 * 1000
            },
            featureConfig: FEATURE_CONFIG.gene
          })
        });
        genomeViewer.addTrack(tracks.icgcGeneTrack);

        var icgcGeneOverviewRenderer = new FeatureRenderer(&#039;gene&#039;);
        icgcGeneOverviewRenderer.on({
          &#039;feature:click&#039;: function(e) {
             scope.$apply(function () {
               $location.path(&#039;/genes/&#039; + e.feature.id);
             });
          }
        });
        tracks.icgcGeneOverviewTrack = new IcgcGeneTrack({
          targetId: null,
          id: 2,
          title: &#039;Gene&#039;,
          histogramZoom: 10,
          labelZoom: 20,
          height: 100,
          visibleRange: {start: 0, end: 100},
          titleVisibility: &#039;hidden&#039;,
          featureTypes: FEATURE_TYPES,

          renderer: icgcGeneOverviewRenderer,

          dataAdapter: new IcgcGeneAdapter({
            category: &quot;genomic&quot;,
            subCategory: &quot;region&quot;,
            resource: &quot;gene&quot;,
            params: {
              exclude: &#039;transcripts&#039;
            },
            species: genomeViewer.species,
            featureCache: {
              gzip: true,
              chunkSize: 50000
            },
            filters: {},
            options: {
              histogramThreshold: 20 * 1000 * 1000,
              transcriptThreshold: 10 * 1000 * 1000
            },
            featureConfig: FEATURE_CONFIG.gene
          })
        });
        genomeViewer.addOverviewTrack(tracks.icgcGeneOverviewTrack);

        tracks.icgcMutationsTrack = new IcgcMutationTrack({
          targetId: null,
          id: 33333,
          title: &#039;ICGC Mutations&#039;,
          histogramZoom: 70,
          labelZoom: 80,
          height: 100,
          visibleRange: {start: 0, end: 100},
          //  fontClass: &#039;ocb-track-title-font-size&#039;,

          renderer: new FeatureRenderer({
            //  fontClass: &#039;ocb-track-feature-font-size&#039;,
            label: function(f) {
              return f.id;
            },
            tooltipTitle: function(f) {
              return &#039;ICGC mutation&#039; + &#039; - &lt;span class=&quot;ok&quot;&gt;&#039; + f.id + &#039;&lt;/span&gt;&#039;;
            },
            tooltipText: function(f) {
              var consequences = GMService.tooltipConsequences(f.consequences);

              return &#039;mutation:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutation + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;reference allele:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.refGenAllele + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;mutation type:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutationType + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;project info:&lt;span class=&quot;ssel&quot;&gt;&lt;br&gt;&#039; + f.projectInfo.join(&#039;&lt;br&gt;&#039;) + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;consequences:&lt;br&gt;&lt;span class=&quot;ssel&quot;&gt;&#039; + consequences + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;source:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;ICGC&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;start-end:&amp;nbsp;&lt;span class=&quot;emph&quot;&gt;&#039; + f.start + &#039;-&#039; + f.end + &#039;&lt;/span&gt;&lt;br&gt;&#039;;
            },
            color: function(feat) {
              switch (feat.mutationType) {
              case &#039;single base substitution&#039;:
                return &#039;Chartreuse&#039;;
              case &#039;insertion of &lt;=200bp&#039;:
                return &#039;orange&#039;;
              case &#039;deletion of &lt;=200bp&#039;:
                return &#039;red&#039;;
              case &#039;multiple base substitution (&gt;=2bp and &lt;=200bp)&#039;:
                return &#039;lightblue&#039;;
              default:
                return &#039;black&#039;;
              }
            },
            infoWidgetId: &#039;mutationCds&#039;,
            height: 8,
            histogramColor: &#039;orange&#039;,
            handlers: {
              &#039;feature:mouseover&#039;: function(e) {
                console.log(e);
              },
              &#039;feature:click&#039;: function(e) {
                scope.$apply(function () {
                  $location.path(&#039;/mutations/&#039; + e.feature.id)
                });
              }
            }
          }),

          dataAdapter: new IcgcMutationAdapter({
            category: &#039;genomic&#039;,
            subCategory: &#039;region&#039;,
            resource: &#039;snp&#039;,
            params: {
              exclude: &#039;transcriptVariations,xrefs,samples&#039;
            },
            species: genomeViewer.species,
            featureCache: {
              gzip: true,
              chunkSize: 10000
            },
            filters: {},
            options: {},
            featureConfig: FEATURE_CONFIG.snp
          })
        });
        genomeViewer.addTrack(tracks.icgcMutationsTrack);

        genomeViewer.karyotypePanel.hide();
//        genomeViewer.karyotypePanel.showContent();
        genomeViewer.chromosomePanel.hide();
//        genomeViewer.chromosomePanel.showContent();
      }

      scope.$watch(&#039;[genes, mutations, tab]&#039;, function(params) {
        var genes, mutations, tab;

        genes = params[0];
        mutations = params[1];
        tab = params[2];

        if (!done) {
          if (tab === &#039;genes&#039; &amp;&amp;
            genes.hasOwnProperty(&#039;data&#039;) &amp;&amp;
            genes.data.hasOwnProperty(&#039;hits&#039;) &amp;&amp;
            genes.data.hits.length) {
            done = true;
            var gene = genes.data.hits[0];
            regionObj = new Region({chromosome: gene.chromosome, start: gene.start - 1500, end: gene.end + 1500});
            setup();
          }
          else if (tab === &#039;mutations&#039; &amp;&amp;
            mutations.hasOwnProperty(&#039;data&#039;) &amp;&amp;
            mutations.data.hasOwnProperty(&#039;hits&#039;) &amp;&amp;
            mutations.data.hits.length) {
            done = true;
            var mutation = mutations.data.hits[0];
            regionObj = new Region({
              chromosome: mutation.chromosome,
              start: mutation.chromosomeStart,
              end: mutation.chromosomeStart
            });
            setup();
          }

        }
      }, true);

      scope.$on(&#039;gv:set:region&#039;, function(e, params) {
        if (genomeViewer) {
          genomeViewer.setRegion(params);
          scope.$broadcast(&#039;gv:zoom:set&#039;, navigationBar._calculateZoomByRegion());
        }
      });
      scope.$on(&#039;gv:toggle:panel&#039;, function(e, params) {
        var action = params.active ? &#039;show&#039; : &#039;hide&#039;;
        genomeViewer[params.panel + &#039;Panel&#039;][action]();
      });
      scope.$on(&#039;gv:zoom:set&#039;, function(e, val) {
        navigationBar._handleZoomSlider(val);
      });
      scope.$on(&#039;gv:reset&#039;, function(e) {
        navigationBar._handleRestoreDefaultRegion(e);
        scope.$broadcast(&#039;gv:zoom:set&#039;, navigationBar._calculateZoomByRegion());
      });
      scope.$on(&#039;gv:autofit&#039;, function() {
        genomeViewer.enableAutoHeight();
      });
    }
  };
});

angular.module(&#039;icgc.modules.genomeviewer&#039;).directive(&#039;gvembed&#039;, function(GMService, $location) {
  return {
    restrict: &#039;E&#039;,
    replace: true,
    transclude: true,
    template: &#039;&lt;div id=&quot;gv-application&quot; style=&quot;&quot;&gt;&#039; + &#039;&lt;/div&gt;&#039;,

    link: function(scope, element, attrs) {
      var genomeViewer, navigationBar, tracks = {};

      function setup(regionObj) {

        genomeViewer = genomeViewer || new GenomeViewer({
          region: regionObj,
          defaultRegion: regionObj,
          sidePanel: false,
          targetId: &#039;gv-application&#039;,
          autoRender: true,
          //trackListTitle: &#039;&#039;,
          drawNavigationBar: false,
          drawKaryotypePanel: false,
          drawChromosomePanel: false,
          drawRegionOverviewPanel: true,
          version: &#039;&lt;small class=&quot;pull-left&quot;&gt;Powered by &#039; +
            &#039;&lt;a target=&quot;_blank&quot; href=&quot;http://www.genomemaps.org/&quot;&gt;Genome Maps&lt;/a&gt;&lt;/small&gt;&#039;
        });

        navigationBar = new IcgcNavigationBar({
          handlers: {
            &#039;region:change&#039;: function(event) {
              Utils.setMinRegion(event.region, genomeViewer.getSVGCanvasWidth());
              genomeViewer.trigger(&#039;region:change&#039;, event);
            },
            &#039;restoreDefaultRegion:click&#039;: function(event) {
              Utils.setMinRegion(genomeViewer.defaultRegion, genomeViewer.getSVGCanvasWidth());
              event.region = genomeViewer.defaultRegion;
              genomeViewer.trigger(&#039;region:change&#039;, event);
            }
          }
        });
        genomeViewer.on(&#039;region:change&#039;, function(event) {
          if (event.sender !== navigationBar) {
            navigationBar.setRegion(event.region);
          }
        });
        genomeViewer.on(&#039;region:move&#039;, function(event) {
          if (event.sender !== navigationBar) {
            navigationBar.moveRegion(event.region);
          }
        });

        genomeViewer.setNavigationBar(navigationBar);
        genomeViewer.draw();

        // Adding tracks requiered
        tracks.sequence = new SequenceTrack({
          targetId: null,
          id: 1,
          title: &#039;Sequence&#039;,
          histogramZoom: 20,
          transcriptZoom: 50,
          height: 30,
          visibleRange: {start: 99, end: 100},
          featureTypes: FEATURE_TYPES,

          renderer: new SequenceRenderer(),

          dataAdapter: new SequenceAdapter({
            category: &#039;genomic&#039;,
            subCategory: &#039;region&#039;,
            resource: &#039;sequence&#039;,
            species: genomeViewer.species,
            featureCache: {
              gzip: true,
              chunkSize: 1000
            }
          })
        });
        genomeViewer.addTrack(tracks.sequence);

        var icgcGeneOverviewRenderer = new FeatureRenderer(&#039;gene&#039;);
        icgcGeneOverviewRenderer.on({
          &#039;feature:click&#039;: function(e) {
            scope.$apply(function () {
              $location.path(&#039;/genes/&#039; + e.feature.id)
            });
          }
        });
        tracks.icgcGeneOverviewTrack = new IcgcGeneTrack({
          targetId: null,
          id: 2,
          title: &#039;Gene&#039;,
          histogramZoom: 10,
          labelZoom: 20,
          height: 100,
          visibleRange: {start: 0, end: 100},
          titleVisibility: &#039;hidden&#039;,
          featureTypes: FEATURE_TYPES,

          renderer: icgcGeneOverviewRenderer,

          dataAdapter: new IcgcGeneAdapter({
            category: &quot;genomic&quot;,
            subCategory: &quot;region&quot;,
            resource: &quot;gene&quot;,
            params: {
              exclude: &#039;transcripts&#039;
            },
            species: genomeViewer.species,
            featureCache: {
              gzip: true,
              chunkSize: 50000
            },
            filters: {},
            options: {
              histogramThreshold: 20 * 1000 * 1000,
              transcriptThreshold: 10 * 1000 * 1000
            },
            featureConfig: FEATURE_CONFIG.gene
          })
        });
        genomeViewer.addOverviewTrack(tracks.icgcGeneOverviewTrack);

        tracks.icgcGeneTrack = new IcgcGeneTrack({
          targetId: null,
          id: 2,
          title: &#039;Genes&#039;,
          histogramZoom: 20,
          transcriptZoom: 40,
          height: 260,
          visibleRange: {start: 0, end: 100},
          featureTypes: FEATURE_TYPES,

          renderer: new GeneRenderer({
            handlers: {
              &#039;feature:click&#039;: function (e) {
                var path = &#039;/genes/&#039; + e.feature[e.featureType === &#039;gene&#039; ? &#039;id&#039; : &#039;geneId&#039;];
                scope.$apply(function () {
                  $location.path(path).search({});
                });
              }
            }
          }),

          dataAdapter: new IcgcGeneAdapter({
            category: &#039;genomic&#039;,
            subCategory: &#039;region&#039;,
            resource: &#039;gene&#039;,
            params: {
              exclude: &#039;transcriptVariations,xrefs,samples&#039;
            },
            species: genomeViewer.species,
            featureCache: {
              gzip: true,
              chunkSize: 50000
            },
            filters: {},
            options: {},
            featureConfig: FEATURE_CONFIG.gene
          })
        });
        genomeViewer.addTrack(tracks.icgcGeneTrack);

        tracks.icgcMutationsTrack = new IcgcMutationTrack({
          targetId: null,
          id: 3,
          title: &#039;ICGC Mutations&#039;,
          histogramZoom: 70,
          labelZoom: 80,
          height: 100,
          visibleRange: {start: 0, end: 100},
          renderer: new FeatureRenderer({
            label: function(f) {
              return f.id;
            },
            tooltipTitle: function(f) {
              return &#039;ICGC mutation&#039; + &#039; - &lt;span class=&quot;ok&quot;&gt;&#039; + f.id + &#039;&lt;/span&gt;&#039;;
            },
            tooltipText: function(f) {
              var consequences = GMService.tooltipConsequences(f.consequences);
              return &#039;mutation:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutation + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;reference allele:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.refGenAllele + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;mutation type:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutationType + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;project info:&lt;span class=&quot;ssel&quot;&gt;&lt;br&gt;&#039; + f.projectInfo.join(&#039;&lt;br&gt;&#039;) + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;consequences:&lt;br&gt;&lt;span class=&quot;ssel&quot;&gt;&#039; + consequences + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;source:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;ICGC&lt;/span&gt;&lt;br&gt;&#039; +
                &#039;start-end:&amp;nbsp;&lt;span class=&quot;emph&quot;&gt;&#039; + f.start + &#039;-&#039; + f.end + &#039;&lt;/span&gt;&lt;br&gt;&#039;;
            },
            color: function(feat) {
              switch (feat.mutationType) {
              case &#039;single base substitution&#039;:
                return &#039;Chartreuse&#039;;
              case &#039;insertion of &lt;=200bp&#039;:
                return &#039;orange&#039;;
              case &#039;deletion of &lt;=200bp&#039;:
                return &#039;red&#039;;
              case &#039;multiple base substitution (&gt;=2bp and &lt;=200bp)&#039;:
                return &#039;lightblue&#039;;
              default:
                return &#039;black&#039;;
              }
            },
            infoWidgetId: &#039;mutationCds&#039;,
            height: 8,
            histogramColor: &#039;orange&#039;,
            handlers: {
              &#039;feature:mouseover&#039;: function(e) {
              },
              &#039;feature:click&#039;: function(e) {
                scope.$apply(function () {
                  $location.path(&#039;/mutations/&#039; + e.feature.id);
                });
              }
            }
          }),

          dataAdapter: new IcgcMutationAdapter({
            category: &#039;genomic&#039;,
            subCategory: &#039;region&#039;,
            resource: &#039;snp&#039;,
            params: {
              exclude: &#039;transcriptVariations,xrefs,samples&#039;
            },
            species: genomeViewer.species,
            featureCache: {
              gzip: true,
              chunkSize: 10000
            },
            filters: {},
            options: {},
            featureConfig: FEATURE_CONFIG.snp
          })
        });
        genomeViewer.addTrack(tracks.icgcMutationsTrack);

        scope.$on(&#039;gv:set:region&#039;, function(e, params) {
          genomeViewer.setRegion(params);
        });
        scope.$on(&#039;gv:zoom:set&#039;, function(e, val) {
          navigationBar._handleZoomSlider(val);
        });
        scope.$on(&#039;gv:reset&#039;, function(e) {
          console.log(&#039;here&#039;);
          navigationBar._handleRestoreDefaultRegion(e);
        });
        scope.$on(&#039;gv:autofit&#039;, function() {
          genomeViewer.enableAutoHeight();
        });
      }

      attrs.$observe(&#039;region&#039;, function(region) {
        if (!region) {
          return;
        }

        var regionObj = new Region();
        regionObj.parse(region);
        var offset = (regionObj.end - regionObj.start)*0.05;
        regionObj.start -= offset;
        regionObj.end += offset;
        setup(regionObj);
      });
    }
  };
});</textarea>
  </div>
</div>

<footer class="footer">
  <div class="container">
    <p>.</p>
  </div>
</footer>

<script type="text/html" id="complexity-popover-template">
  <div class="complexity-notice">
    Complexity : {{ complexity.cyclomatic }} <br>
    Length : {{ complexity.halstead.length }} <br>
    Difficulty : {{ complexity.halstead.difficulty.toFixed(2) }} <br>
    Est # bugs : {{ complexity.halstead.bugs.toFixed(2) }}<br>
  </div>
</script>

<script type="text/javascript" src="../../assets/scripts/bundles/core-bundle.js"></script>
<script type="text/javascript" src="../../assets/scripts/bundles/codemirror.js"></script>
<script type="text/javascript" src="../../assets/scripts/codemirror.markpopovertext.js"></script>
<script type="text/javascript" src="report.js"></script>
<script type="text/javascript" src="report.history.js"></script>
<script type="text/javascript" src="../../assets/scripts/plato-file.js"></script>
</body>
</html>
