<h1>
    <img src="styles/images/PCAWG-icon.png" style="height:3.5rem; margin-left:2rem; vertical-align:bottom"></img> 
    <span style="display:inline-block">PCAWG - Pancancer Analysis of Whole Genomes</span>
</h1>
<section style="padding-top:0"> 
   <div class="half pull-right" style="margin-bottom:1.5rem; margin-top:1rem; margin-left:2rem; margin-right: 2rem">

      <stacked height="250"
         data-title="Donor Distribution by Primary Site"
         data-subtitle="{{summary.projectCount}} projects and {{summary.primarySiteCount}} primary sites"
         data-y-label="Donors"
         data-alternate-brightness="true"
         data-items="primarySites">
      </stacked>

      <section class="well" style="padding-top:0.5rem; padding-bottom:0.5rem">
          <div class="third text-center">
              <i class="icon-team" style="font-size:1.25rem"></i><strong>{{summary.donorCount | number}}</strong> Donors
          </div>
          <div class="third text-center">
              <i class="icon-file" style="font-size:1.25rem"></i><strong>{{summary.fileCount | number}}</strong> Files
          </div>
          <div class="third omega text-center">
              <i class="icon-database" style="font-size:1.25rem"></i><strong>{{summary.totalFileSize | bytes}}</strong>
          </div>
      </section>

      <table class="table table-info">
          <thead>
              <th>Data Type</th>
              <th># Donors</th>
              <th># Files</th>
              <th>Format</th>
              <th>Size</th>
          </thead>
          <tbody>
              <tr data-ng-repeat="type in pcawgDatatypes">
                  <td> {{type.uiName}}</td>
                  <td class="text-right">{{type.donorCount | number}}</td>
                  <td class="text-right">
                      <a href="/repository/external?filters={{type.filters}}">{{type.fileCount | number}}</a>
                  </td>
                  <td>{{ type.fileFormat.join(', ')}}</td>
                  <td class="text-right">{{type.fileSize | bytes}}</td>
              </tr>
          </tbody>
      </table>

      <span>
          Available data as of <a href="/repository/external?filters={{filters}}">{{indexDate | date}}</a>
      </span> 
   </div>


   <p style="text-align:justify; text-indent:1rem">
       The Pancancer Analysis of Whole Genomes (PCAWG) study is an international collaboration to identify common patterns of mutation in more than 2,600 cancer whole genomes from 
       the International Cancer Genome Consortium. Building upon previous work which examined cancer coding regions (Cancer Genome Atlas Research Network, The Cancer Genome Atlas 
       Pan-Cancer analysis project, 
       <a href="http://www.nature.com/ng/journal/v45/n10/full/ng.2764.html" target="_blank"><i class="icon-external-link"></i>&nbsp;&nbsp;Nat. Genet. 2013 45:1113</a>), 
       this project is exploring the nature and consequences of somatic and germline variations in both coding and non-coding regions, with specific emphasis on cis-regulatory sites, 
       non-coding RNAs, and large-scale structural alterations.
   </p>
   <p style="text-align:justify; text-indent:1rem">
       In order to facilitate the comparison among diverse tumor types, all tumor and matched normal genomes have been subjected to a uniform set of alignment and variant calling algorithms, and must pass a rigorous set of qualilty control tests. The research activities are coordinated by a series of working groups comprising more than 700 scientists and covering the following themes:
   </p>
   <p>
       <ol>
           <li>Novel somatic mutation calling methods</li>
           <li>Analysis of mutations in regulatory regions</li>
           <li>Integration of the transcriptome and genome</li>
           <li>Integration of the epigenome and genome</li>
           <li>Consequences of somatic mutations on pathway and network activity</li>
           <li>Patterns of structural variations, signatures, genomic correlations, retrotransposons and mobile elements</li>
           <li>Mutation signatures and processes</li>
           <li>Germline cancer genome</li>
           <li>Inferring driver mutations and identifying cancer genes and pathways</li>
           <li>Translating cancer genomes to the clinic</li>
           <li>Evolution and heterogeneity</li>
           <li>Portals, visualization and software infrastructure</li>
           <li>Molecular subtypes and classification</li>
           <li>Analysis of mutations in non-coding RNA</li>
           <li>Mitochondrial</li>
           <li>Pathogens</li>
       </ol>

       All data resulting from this study will be released to the public under the terms of the
       <a href="http://www.icgc.org/" target="_blank"><i class="icon-external-link"></i>ICGC</a> and
       <a href="http://cancergenome.nih.gov/" target="_blank"><i class="icon-external-link"></i>TCGA</a> 
       data release policies. Please contact the <a href="mailto:dcc-support@icgc.org"><i class="icon-mail"></i>ICGC helpdesk</a> for more information.
   </p>


</section>

