<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title></title>

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="description" content="">
  <meta name="author" content="">

  <!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->

  <link href="../../assets/css/vendor/morris.css" rel="stylesheet">
  <link href="../../assets/css/vendor/bootstrap-3.0.0-wip.css" rel="stylesheet">
  <link href="../../assets/css/vendor/font-awesome.css" rel="stylesheet">
  <link href="../../assets/css/vendor/font-awesome-ie7.css" rel="stylesheet">
  <link href="../../assets/css/vendor/codemirror.css" rel="stylesheet">
  <link href="../../assets/css/plato.css" rel="stylesheet">
  <link href="../../assets/css/plato-file.css" rel="stylesheet">

</head>

<body>

<div class="navbar navbar-fixed-top">
  <div class="container">
    <a class="brand" href="http://github.com/jsoverson/plato">Plato on Github</a>
    <ul class="nav">
      <li>
        <a href="../../index.html">Report Home</a>
      </li>
    </ul>
  </div>
</div>

<div class="jumbotron">
  <div class="container">
    <h1>app/scripts/modules/genomemaps/gmdirective.js</h1>
  </div>
</div>

<div class="container aggregate-stats">
  <div class="row">
    <div class="span6">
      <h2 class="header">Maintainability <a href="http://blogs.msdn.com/b/codeanalysis/archive/2007/11/20/maintainability-index-range-and-meaning.aspx"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="A value between 0 and 100 that represents the relative ease of maintaining the code. A high value means better maintainability." data-original-title="Maintainability Index"></i></a></h2>
      <p class="stat">63.19</p>
    </div>
    <div class="span6">
      <h2 class="header">Lines of code <i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Source Lines of Code / Logical Lines of Code" data-original-title="SLOC/LSLOC"></i></h2>
      <p class="stat">920</p>
    </div>
  </div>
  <div class="row historical">
    <div class="span6">
      <p id="chart_historical_maint" class="chart"></p>
    </div>
    <div class="span6">
      <p id="chart_historical_sloc" class="chart"></p>
    </div>
  </div>
  <div class="row">
    <div class="span6">
      <h2 class="header">Difficulty  <a href="http://en.wikipedia.org/wiki/Halstead_complexity_measures"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="The difficulty measure is related to the difficulty of the program to write or understand." data-original-title="Difficulty"></i></a></h2>
      <p class="stat">79.44</p>
    </div>
    <div class="span6">
      <h2 class="header">Estimated Errors  <a href="http://en.wikipedia.org/wiki/Halstead_complexity_measures"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Halstead's delivered bugs is an estimate for the number of errors in the implementation." data-original-title="Delivered Bugs"></i></a></h2>
      <p class="stat">8.94</p>
    </div>
  </div>
</div>

<div class="container charts">
  <div class="row">
    <h2 class="header">Function weight</h2>
  </div>
  <div class="row">
    <div class="span6">
      <h3 class="chart-header">By Complexity <a href="http://en.wikipedia.org/wiki/Cyclomatic_complexity"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="This metric counts the number of distinct paths through a block of code. Lower values are better." data-original-title="Cyclomatic Complexity"></i></a></h3>
      <div id="fn-by-complexity" class="stat"></div>
    </div>
    <div class="span6">
      <h3 class="chart-header">By SLOC  <i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Source Lines of Code / Logical Lines of Code" data-original-title="SLOC/LSLOC"></i></h3>
      <div id="fn-by-sloc" class="stat"></div>
    </div>
  </div>
</div>

<div class="container">
  <div class="row">
    <textarea id="file-source" class="span12">/*
 * Copyright 2013(c) The Ontario Institute for Cancer Research. All rights reserved.
 *
 * This program and the accompanying materials are made available under the terms of the GNU Public
 * License v3.0. You should have received a copy of the GNU General Public License along with this
 * program. If not, see &lt;http://www.gnu.org/licenses/&gt;.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS &quot;AS IS&quot; AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


var mutationGenomeViewer, advSearchGeneGenomeViewer, advSearchMutationGenomeViewer;

angular.module(&#039;gm&#039;, [&#039;gm.directives&#039;]);

angular.module(&#039;gm.directives&#039;, []);

angular.module(&quot;gm.directives&quot;).directive(&#039;gvhover&#039;, function ($window) {

    return {
        restrict: &#039;E&#039;,
        scope: {

            id: &quot;=&quot;,
            chromosome: &quot;=&quot;,
            start: &quot;=&quot;
        },

        replace: true,
        template: &#039;&lt;a href=&quot;#/mutations/{{id}}&quot;&gt;{{id}}&lt;/a&gt;&#039;,
        link: function (scope, elem, attrs) {
//            $window.mutationGenomeViewer.hightlight({attrValues: scope.id});

//            mutationGenomeViewer.highlight({attrValues: scope.id});
//            gm.getTrackByName().addFilter();
//            (gm.draw();)
            scope.highlight = function () {
                console.log(scope.chromosome + &quot;:&quot; + scope.start + &quot;-&quot; + scope.start);

                var disp = parseInt(scope.start) - mutationGenomeViewer.region.center();
                console.log(&#039;disp: &#039; + disp);
//                var region = new Region();
//                region.load({chromosome: scope.chromosome, start: parseInt(scope.start)+disp, end: parseInt(scope.start)+disp});

                mutationGenomeViewer.move(disp);

                console.log(&#039;GV region center: &#039; + mutationGenomeViewer.region.center());
                mutationGenomeViewer.highlight({attrName: &#039;feature_id&#039;, attrValue: scope.id});
            }
        }
    }
});

angular.module(&quot;gm.directives&quot;).directive(&#039;gvgenehover&#039;, function ($window) {

    return {
        restrict: &#039;E&#039;,
        scope: {
            id: &quot;=&quot;,
            symbol: &quot;=&quot;,
            chromosome: &quot;=&quot;,
            start: &quot;=&quot;,
            end: &quot;=&quot;
        },

        replace: true,
        template: &#039;&lt;span&gt;&amp;nbsp;&lt;img id=&quot;gene-view-{{id}}&quot; src=&quot;styles/images/eye_icon.png&quot; style=&quot;float:right&quot; /&gt;&lt;/span&gt;&#039;,
//    &lt;a href=&quot;#/genes/{{id}}&quot;&gt;{{symbol}}&lt;/a&gt;
        link: function (scope, elem, attrs) {
            scope.viewGene = function () {
                console.log(&quot;gv-gene-hover: &quot; + scope.chromosome + &quot;:&quot; + scope.start + &quot;-&quot; + scope.end);
                console.log({chromosome: scope.chromosome, start: scope.start, end: scope.end});

                advSearchGeneGenomeViewer.setRegion({chromosome: scope.chromosome, start: scope.start - 1500, end: scope.end + 1500});
            }
        }
    }
});

angular.module(&quot;gm.directives&quot;).directive(&#039;gvmutationhover&#039;, function ($window) {

    return {
        restrict: &#039;E&#039;,
        scope: {
            id: &quot;=&quot;,
            chromosome: &quot;=&quot;,
            start: &quot;=&quot;,
            end: &quot;=&quot;
        },

        replace: true,
        template: &#039;&lt;span&gt;&amp;nbsp;&lt;img id=&quot;mutation-view-{{id}}&quot; src=&quot;styles/images/eye_icon.png&quot; style=&quot;float:right&quot; /&gt;&lt;/span&gt;&#039;,
//    &lt;a href=&quot;#/genes/{{id}}&quot;&gt;{{symbol}}&lt;/a&gt;
        link: function (scope, elem, attrs) {
            scope.viewMutation = function () {
                console.log(&quot;gv-mutation-hover: &quot; + scope.chromosome + &quot;:&quot; + scope.start + &quot;-&quot; + scope.end);
//                var disp = parseInt(scope.start) - mutationGenomeViewer.region.center();
//                console.log(&#039;disp: &#039; + disp);
//                mutationGenomeViewer.move(disp);
//                console.log(&#039;GV region center: &#039; + mutationGenomeViewer.region.center());
//                Region reg = new Region({chromosome: scope.chromosome, start: scope.start, end: scope.end});
                console.log({chromosome: scope.chromosome, start: scope.start, end: scope.end});

                advSearchMutationGenomeViewer.setRegion({chromosome: scope.chromosome, start: scope.start, end: scope.end});
            }
        }
    }
});

angular.module(&#039;gm.directives&#039;)
    .directive(&#039;gm&#039;, function ($location) {
        return {
            restrict: &#039;E&#039;,
            replace: true,
            transclude: true,
            scope: {},
            template: &#039;&lt;div id=&quot;gm&quot;&gt;&#039; +
                &#039;&lt;/div&gt;&#039;,

            link: function (scope, element, attrs) {

            }
        };
    });

angular.module(&#039;gm.directives&#039;)
    .directive(&#039;gmembed&#039;, function () {
        return {
            restrict: &#039;E&#039;,
            replace: true,
            transclude: true,
            scope: {region: &#039;@region&#039;},
            template: &#039;&lt;div id=&quot;gv-application&quot; style=&quot;&quot;&gt;&#039; + &#039;&lt;/div&gt;&#039;,

            link: function (scope, element, attrs) {

                attrs.$observe(&#039;region&#039;, function (region) {
                    if (region) {
                        console.log(region);
                        var regionObj = new Region();
                        regionObj.parse(region);
                        console.log(regionObj);

                        $(document).ready(function () {
                            window.projectFilterValues = [];
                            window.checkValues = [];
                            window.virtualTrackFilters = [];
                            mutationGenomeViewer = new GenomeViewer({
                                width: $(&quot;.col-span-10&quot;).outerWidth(),
                                height: 580,
                                region: regionObj,
                                defaultRegion: regionObj,
                                sidePanel: false,
                                targetId: &#039;gv-application&#039;,
                                autoRender: true,
                                trackListTitle: &#039;&#039;,
                                drawNavigationBar: false,
                                drawKaryotypePanel: false,
                                drawChromosomePanel: false,
                                drawRegionOverviewPanel: false,
                                version: &#039;&#039;
                            });

                            mutationGenomeViewer.draw();

                            var navigationBar = new IcgcNavigationBar({
                                handlers: {
                                    &#039;region:change&#039;: function (event) {
                                        Utils.setMinRegion(event.region, mutationGenomeViewer.getSVGCanvasWidth())
                                        mutationGenomeViewer.trigger(&#039;region:change&#039;, event);
                                    },
                                    &#039;size:small&#039;: function (event) {
                                        mutationGenomeViewer.setWidth(parseInt($(&#039;#gv-application&#039;).parent().width()));
                                    },
                                    &#039;size:big&#039;: function (event) {
                                        mutationGenomeViewer.setWidth(parseInt($(window).width() - 30));
                                    },
//                                    &#039;fullscreen:click&#039;: function (event) {
//                                        if (mutationGenomeViewer.fullscreen) {
//                                            $(mutationGenomeViewer.div).css({width:&#039;auto&#039;});
//                                            Utils.cancelFullscreen();//no need to pass the dom object;
//                                            mutationGenomeViewer.fullscreen = false;
//                                        } else {
//                                            $(mutationGenomeViewer.div).css({width:screen.width});
//                                            Utils.launchFullScreen(mutationGenomeViewer.div);
//                                            mutationGenomeViewer.fullscreen = true;
//                                        }
//                                    },
                                    &#039;restoreDefaultRegion:click&#039;: function (event) {
                                        Utils.setMinRegion(mutationGenomeViewer.defaultRegion, mutationGenomeViewer.getSVGCanvasWidth());
                                        event.region = mutationGenomeViewer.defaultRegion;
                                        mutationGenomeViewer.trigger(&#039;region:change&#039;, event);
                                    }
                                }
                            });
                            mutationGenomeViewer.on(&#039;region:change&#039;, function (event) {
                                if (event.sender != navigationBar) {
                                    navigationBar.setRegion(event.region);
                                }
                            });
                            mutationGenomeViewer.on(&#039;region:move&#039;, function (event) {
                                if (event.sender != navigationBar) {
                                    navigationBar.moveRegion(event.region);
                                }
                            });
                            mutationGenomeViewer.on(&#039;width:change&#039;, function (event) {
                                navigationBar.setWidth(event.width);
                            });



                            mutationGenomeViewer.setNavigationBar(navigationBar);
                            navigationBar.draw();

//                    // Making some panel not visible
//                    mutationGenomeViewer.navigationBar.setVisible({&#039;species&#039;:false, &#039;chromosomes&#039;: false, &#039;location&#039;: false, &#039;movement&#039;: false, &#039;search&#039;: false});
//                            mutationGenomeViewer.setKaryotypePanelVisible(false);
//                            mutationGenomeViewer.setChromosomePanelVisible(false);
//                            mutationGenomeViewer.setRegionOverviewPanelVisible(false);

                            // Adding tracks requiered
                            this.sequence = new SequenceTrack({
                                targetId: null,
                                id: 1,
                                title: &#039;Sequence&#039;,
                                histogramZoom: 20,
                                transcriptZoom: 50,
                                height: 30,
                                visibleRange: {start: 100, end: 100},
                                featureTypes: FEATURE_TYPES,

                                renderer: new SequenceRenderer(),

                                dataAdapter: new SequenceAdapter({
                                    category: &quot;genomic&quot;,
                                    subCategory: &quot;region&quot;,
                                    resource: &quot;sequence&quot;,
                                    species: mutationGenomeViewer.species,
                                    featureCache: {
                                        gzip: true,
                                        chunkSize: 1000
                                    }
                                })
                            });

                            mutationGenomeViewer.addTrack(this.sequence);

                            this.icgcGeneTrack = new IcgcGeneTrack({
                                targetId: null,
                                id: 2,
                                title: &#039;ICGC Genes&#039;,
                                histogramZoom: 20,
                                transcriptZoom: 50,
                                height: 260,
                                visibleRange: {start: 0, end: 100},
                                featureTypes: FEATURE_TYPES,

                                renderer: new GeneRenderer(),

                                dataAdapter: new IcgcGeneAdapter({
                                    category: &quot;genomic&quot;,
                                    subCategory: &quot;region&quot;,
                                    resource: &quot;snp&quot;,
                                    params: {
                                        exclude: &#039;transcriptVariations,xrefs,samples&#039;
                                    },
                                    species: mutationGenomeViewer.species,
                                    featureCache: {
                                        gzip: true,
                                        chunkSize: 50000
                                    },
                                    filters: {},
                                    options: {},
                                    featureConfig: FEATURE_CONFIG.gene
                                })
                            });
                            mutationGenomeViewer.addTrack(this.icgcGeneTrack);


                            this.icgcMutationsTrack = new IcgcMutationTrack({
                                targetId: null,
                                id: 3,
                                title: &#039;ICGC Mutations&#039;,
                                histogramZoom: 70,
                                labelZoom: 80,
                                height: 240,
                                visibleRange: {start: 0, end: 100},
//                        featureTypes:FEATURE_TYPES,

                                renderer: new FeatureRenderer({
                                    label: function (f) {
                                        return f.id;
                                    },
                                    tooltipTitle: function (f) {
                                        return &#039;ICGC mutation&#039; + &#039; - &lt;span class=&quot;ok&quot;&gt;&#039; + f.id + &#039;&lt;/span&gt;&#039;;
                                    },
                                    tooltipText: function (f) {
                                        var consequences = &#039;&#039;;
                                        for (var i = 0; i &lt; f.consequences.length; i++) {
                                            consequences += &#039;&amp;nbsp;&#039; + f.consequences[i] + &#039;&lt;br&gt;&#039;;
                                        }
                                        return &#039;mutation:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutation + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;reference allele:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.refGenAllele + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;mutation type:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutationType + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;project info:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.projectInfo[0] + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;consequences:&lt;br&gt;&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + consequences + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;source:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;ICGC&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;start-end:&amp;nbsp;&lt;span class=&quot;emph&quot;&gt;&#039; + f.start + &#039;-&#039; + f.end + &#039;&lt;/span&gt;&lt;br&gt;&#039;;
                                    },
                                    color: function (feat) {
                                        switch (feat.mutationType) {
                                            case &#039;single base substitution&#039;:
                                                return &#039;Chartreuse&#039;;
                                            case &#039;insertion of &lt;=200bp&#039;:
                                                return &#039;orange&#039;;
                                            case &#039;deletion of &lt;=200bp&#039;:
                                                return &#039;red&#039;;
                                            case &#039;multiple base substitution (&gt;=2bp and &lt;=200bp)&#039;:
                                                return &#039;lightblue&#039;;
                                            default:
                                                return &#039;black&#039;;
                                        }
                                    },
                                    infoWidgetId: &quot;mutationCds&quot;,
                                    height: 8,
                                    histogramColor: &quot;orange&quot;,
                                    handlers: {
                                        &#039;feature:mouseover&#039;: function (e) {
//                                    console.log(e)
                                        },
                                        &#039;feature:click&#039;: function (e) {
                                            console.log(e)
                                        }
                                    }
                                }),

                                dataAdapter: new IcgcMutationAdapter({
                                    category: &quot;genomic&quot;,
                                    subCategory: &quot;region&quot;,
                                    resource: &quot;snp&quot;,
                                    params: {
                                        exclude: &#039;transcriptVariations,xrefs,samples&#039;
                                    },
                                    species: mutationGenomeViewer.species,
                                    featureCache: {
                                        gzip: true,
                                        chunkSize: 10000
                                    },
                                    filters: {},
                                    options: {},
                                    featureConfig: FEATURE_CONFIG.snp
                                })
                            });
                            mutationGenomeViewer.addTrack(this.icgcMutationsTrack);

                        });
                    }
                });
            }
        };
    });

angular.module(&#039;gm.directives&#039;)
    .directive(&#039;gmadvsearchgeneembed&#039;, function () {
        return {
            restrict: &#039;E&#039;,
            replace: true,
            transclude: true,
            scope: {genes: &#039;@genes&#039;},
            template: &#039;&lt;div id=&quot;gv-adv-search-gene&quot; style=&quot;float:right;width: 900px; margin-left: 0px; z-index:1&quot;&gt;&#039; + &#039;&lt;/div&gt;&#039;,

            link: function (scope, element, attrs) {

                attrs.$observe(&#039;genes&#039;, function (genes) {

//                    if(advSearchGeneGenomeViewer != null) {
//                        return;
//                    }
                    if (genes) {
                        console.log(genes);
                        var genesObj = JSON.parse(genes);
                        var gene = genesObj[0];
                        console.log(gene);

                        var regionObj = new Region({chromosome: gene.chromosome, start: gene.start - 1500, end: gene.end + 1500});

//                        regionObj.parse(region);
                        console.log(regionObj);

                        $(document).ready(function () {


                            window.projectFilterValues = [];
                            window.checkValues = [];
                            window.virtualTrackFilters = [];

                            advSearchGeneGenomeViewer = new GenomeViewer({
                                width: $(&quot;.col-span-9&quot;).outerWidth(),
                                height: 580,
                                region: regionObj,
                                defaultRegion: regionObj,
                                sidePanel: false,
                                targetId: &#039;gv-adv-search-gene&#039;,
                                autoRender: true,
                                trackListTitle: &#039;&#039;,
                                drawNavigationBar: false,
                                drawKaryotypePanel: false,
                                drawChromosomePanel: false,
                                drawRegionOverviewPanel: false,
                                version: &#039;&#039;
                            });

                            advSearchGeneGenomeViewer.draw();

                            var navigationBar = new IcgcNavigationBar({
                                handlers: {
                                    &#039;region:change&#039;: function (event) {
                                        Utils.setMinRegion(event.region, advSearchGeneGenomeViewer.getSVGCanvasWidth())
                                        advSearchGeneGenomeViewer.trigger(&#039;region:change&#039;, event);
                                    },
                                    &#039;size:small&#039;: function (event) {
                                        advSearchGeneGenomeViewer.setWidth(parseInt($(&#039;#gv-adv-search-gene&#039;).parent().width()));
                                    },
                                    &#039;size:big&#039;: function (event) {
                                        advSearchGeneGenomeViewer.setWidth(parseInt($(window).width() - 30));
                                    },
//                                    &#039;fullscreen:click&#039;: function (event) {
//                                        if (advSearchGeneGenomeViewer.fullscreen) {
//                                            $(advSearchGeneGenomeViewer.div).css({width:&#039;auto&#039;});
//                                            Utils.cancelFullscreen();//no need to pass the dom object;
//                                            advSearchGeneGenomeViewer.fullscreen = false;
//                                        } else {
//                                            $(advSearchGeneGenomeViewer.div).css({width:screen.width});
//                                            Utils.launchFullScreen(advSearchGeneGenomeViewer.div);
//                                            advSearchGeneGenomeViewer.fullscreen = true;
//                                        }
//                                    },
                                    &#039;restoreDefaultRegion:click&#039;: function (event) {
                                        Utils.setMinRegion(advSearchGeneGenomeViewer.defaultRegion, advSearchGeneGenomeViewer.getSVGCanvasWidth());
                                        event.region = advSearchGeneGenomeViewer.defaultRegion;
                                        advSearchGeneGenomeViewer.trigger(&#039;region:change&#039;, event);
                                    }
                                }
                            });
                            advSearchGeneGenomeViewer.on(&#039;region:change&#039;, function (event) {
                                if (event.sender != navigationBar) {
                                    navigationBar.setRegion(event.region);
                                }
                            });
                            advSearchGeneGenomeViewer.on(&#039;region:move&#039;, function (event) {
                                if (event.sender != navigationBar) {
                                    navigationBar.moveRegion(event.region);
                                }
                            });
                            advSearchGeneGenomeViewer.on(&#039;width:change&#039;, function (event) {
                                navigationBar.setWidth(event.width);
                            });


                            advSearchGeneGenomeViewer.setNavigationBar(navigationBar);
                            navigationBar.draw();

//                    // Making some panel not visible
//                    advSearchGeneGenomeViewer.navigationBar.setVisible({&#039;species&#039;:false, &#039;chromosomes&#039;: false, &#039;location&#039;: false, &#039;movement&#039;: false, &#039;search&#039;: false});
//                            advSearchGeneGenomeViewer.setKaryotypePanelVisible(false);
//                            advSearchGeneGenomeViewer.setChromosomePanelVisible(false);
//                            advSearchGeneGenomeViewer.setRegionOverviewPanelVisible(false);

                            // Adding tracks requiered
                            this.sequence = new SequenceTrack({
                                targetId: null,
                                id: 1,
                                title: &#039;Sequence&#039;,
                                histogramZoom: 20,
                                transcriptZoom: 50,
                                height: 30,
                                visibleRange: {start: 100, end: 100},
                                featureTypes: FEATURE_TYPES,

                                renderer: new SequenceRenderer(),

                                dataAdapter: new SequenceAdapter({
                                    category: &quot;genomic&quot;,
                                    subCategory: &quot;region&quot;,
                                    resource: &quot;sequence&quot;,
                                    species: advSearchGeneGenomeViewer.species,
                                    featureCache: {
                                        gzip: true,
                                        chunkSize: 1000
                                    }
                                })
                            });

                            advSearchGeneGenomeViewer.addTrack(this.sequence);

                            this.icgcGeneTrack = new IcgcGeneTrack({
                                targetId: null,
                                id: 2,
                                title: &#039;ICGC Genes&#039;,
                                histogramZoom: 20,
                                transcriptZoom: 50,
                                height: 260,
                                visibleRange: {start: 0, end: 100},
                                featureTypes: FEATURE_TYPES,

                                renderer: new GeneRenderer(),

                                dataAdapter: new IcgcGeneAdapter({
                                    category: &quot;genomic&quot;,
                                    subCategory: &quot;region&quot;,
                                    resource: &quot;snp&quot;,
                                    params: {
                                        exclude: &#039;transcriptVariations,xrefs,samples&#039;
                                    },
                                    species: advSearchGeneGenomeViewer.species,
                                    featureCache: {
                                        gzip: true,
                                        chunkSize: 10000
                                    },
                                    filters: {},
                                    options: {},
                                    featureConfig: FEATURE_CONFIG.gene
                                })
                            });
                            advSearchGeneGenomeViewer.addTrack(this.icgcGeneTrack);


                            this.icgcMutationsTrack = new IcgcMutationTrack({
                                targetId: null,
                                id: 3,
                                title: &#039;ICGC Mutations&#039;,
                                histogramZoom: 70,
                                labelZoom: 80,
                                height: 240,
                                visibleRange: {start: 0, end: 100},
//                        featureTypes:FEATURE_TYPES,

                                renderer: new FeatureRenderer({
                                    label: function (f) {
                                        return f.id;
                                    },
                                    tooltipTitle: function (f) {
                                        return &#039;ICGC mutation&#039; + &#039; - &lt;span class=&quot;ok&quot;&gt;&#039; + f.id + &#039;&lt;/span&gt;&#039;;
                                    },
                                    tooltipText: function (f) {
                                        var consequences = &#039;&#039;;
                                        for (var i = 0; i &lt; f.consequences.length; i++) {
                                            consequences += &#039;&amp;nbsp;&#039; + f.consequences[i] + &#039;&lt;br&gt;&#039;;
                                        }
                                        return &#039;mutation:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutation + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;reference allele:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.refGenAllele + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;mutation type:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutationType + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;project info:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.projectInfo[0] + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;consequences:&lt;br&gt;&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + consequences + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;source:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;ICGC&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;start-end:&amp;nbsp;&lt;span class=&quot;emph&quot;&gt;&#039; + f.start + &#039;-&#039; + f.end + &#039;&lt;/span&gt;&lt;br&gt;&#039;;
                                    },
                                    color: function (feat) {
                                        switch (feat.mutationType) {
                                            case &#039;single base substitution&#039;:
                                                return &#039;Chartreuse&#039;;
                                            case &#039;insertion of &lt;=200bp&#039;:
                                                return &#039;orange&#039;;
                                            case &#039;deletion of &lt;=200bp&#039;:
                                                return &#039;red&#039;;
                                            case &#039;multiple base substitution (&gt;=2bp and &lt;=200bp)&#039;:
                                                return &#039;lightblue&#039;;
                                            default:
                                                return &#039;black&#039;;
                                        }
                                    },
                                    infoWidgetId: &quot;mutationCds&quot;,
                                    height: 8,
                                    histogramColor: &quot;orange&quot;,
                                    handlers: {
                                        &#039;feature:mouseover&#039;: function (e) {
//                                    console.log(e)
                                        },
                                        &#039;feature:click&#039;: function (e) {
                                            console.log(e)
                                        }
                                    }
                                }),

                                dataAdapter: new IcgcMutationAdapter({
                                    category: &quot;genomic&quot;,
                                    subCategory: &quot;region&quot;,
                                    resource: &quot;snp&quot;,
                                    params: {
                                        exclude: &#039;transcriptVariations,xrefs,samples&#039;
                                    },
                                    species: advSearchGeneGenomeViewer.species,
                                    featureCache: {
                                        gzip: true,
                                        chunkSize: 10000
                                    },
                                    filters: {},
                                    options: {},
                                    featureConfig: FEATURE_CONFIG.snp
                                })
                            });
                            advSearchGeneGenomeViewer.addTrack(this.icgcMutationsTrack);

                        });
                    }
                });
            }
        };
    });

var gvSearchMutationFilter;
angular.module(&#039;gm.directives&#039;)
    .directive(&#039;gmadvsearchmutationembed&#039;, function () {
        return {
            restrict: &#039;E&#039;,
            replace: true,
            transclude: true,
            scope: {mutations: &#039;@mutations&#039;, gvfilters: &#039;@gvfilters&#039;},
            template: &#039;&lt;div style=&quot;width: 900px; margin-left: 20px;&quot;&gt;&lt;div id=&quot;gv-adv-search-table&quot;  style=&quot;float:left;width:150px;height:740px;border:1px solid lightgray;&quot;&gt;&lt;/div&gt;&lt;div id=&quot;gv-adv-search-mutation&quot; style=&quot;width: 700px;float:left&quot;&gt;&#039; + &#039;&lt;/div&gt;&lt;/div&gt;&#039;,

            link: function (scope, element, attrs) {
                attrs.$observe(&#039;gvfilters&#039;, function (gvfilters) {
                    if(gvfilters != &#039;&#039;){
                        gvSearchMutationFilter = JSON.parse(gvfilters);
                        if (advSearchMutationGenomeViewer != null) {
                            advSearchMutationGenomeViewer.trackListPanel.trackSvgList[2].draw();
                        }
                    }
                });
                attrs.$observe(&#039;mutations&#039;, function (mutations) {

                    if (mutations) {
//                        console.log(mutations);
                        var mutationsArray = JSON.parse(mutations);
                        var mutation = mutationsArray[0];

                        console.log(mutation);
                        var regionObj = new Region({chromosome: mutation.chromosome, start: parseInt(mutation.chromosomeStart), end: parseInt(mutation.chromosomeStart)});

//                        regionObj.parse(region);
                        console.log(regionObj);

                        $(document).ready(function () {

                            $(&#039;#gv-adv-search-table&#039;).empty();

                            $(&#039;#gv-adv-search-table&#039;).append(&#039;&lt;div class=&quot;gv-navigation-bar&quot; style=&quot;border-bottom: 1px solid lightgray;&quot;&gt;&#039; +
                                &#039;&lt;a id=&quot;prevMutationPage&quot; class=&quot;btn btn-mini btn-primary&quot;&gt;&lt;i class=&quot;icon-chevron-left&quot;&gt;&lt;/i&gt;&lt;/a&gt; &#039; +
                                &#039;&lt;a id=&quot;nextMutationPage&quot; class=&quot;btn btn-mini btn-primary&quot;&gt;&lt;i class=&quot;icon-chevron-right&quot;&gt;&lt;/i&gt;&lt;/a&gt;&#039; +
                                &#039;&lt;/div&gt;&#039;);

                            if(advSearchMutationGenomeViewer == null) {
                                from = 1;
                            }
                            var prevPage = $(&#039;#gv-adv-search-table&#039;).find(&#039;#prevMutationPage&#039;);
                            $(prevPage).click(function () {
                                if(from &gt; 1){
                                    from -= 10;
                                    window.location.href = &#039;http://localhost:9001/#/search/m?mutations={&quot;from&quot;:&#039;+from+&#039;}&#039;;
                                }
                            });
                            var nextPage = $(&#039;#gv-adv-search-table&#039;).find(&#039;#nextMutationPage&#039;);
                            $(nextPage).click(function () {
                                from += 10;
                                window.location.href = &#039;http://localhost:9001/#/search/m?mutations={&quot;from&quot;:&#039;+from+&#039;}&#039;;
                            });

                            var mDiv = function(mutation){
                                var img = $(&#039;&lt;img src=&quot;styles/images/eye_icon.png&quot; style=&quot;margin: 5px;&quot;/&gt;&#039;)[0];
                                $(img).click(function(){
                                    var region = new Region({chromosome:mutation.chromosome,start:mutation.chromosomeStart,end:mutation.chromosomeStart})
                                    advSearchMutationGenomeViewer.setRegion(region);
                                    console.log(mutation)
                                });
                                var div = $(&#039;&lt;div style=&quot;padding:10px 0 0 10px;&quot;&gt;&lt;a href =&quot;#/mutations/&#039;+mutation.id+&#039;&quot; style=&quot;width:80px;display:inline-block;&quot;&gt;&#039;+ mutation.id+&#039;&lt;/a&gt;&lt;/div&gt;&#039;)[0];

                                var consequences = &#039;&#039;;
                                for (var i =0; i&lt; mutation.consequences.length; i++){
                                    var con = mutation.consequences[i];
                                    var geneSymbol = (con.geneSymbol) ? con.geneSymbol : &#039;&#039;;
                                    var aamuta = (con.aaMutation) ? con.aaMutation : &#039;&#039;;
                                    consequences+=con.consequenceType+&#039;: &#039;+geneSymbol+&#039; &#039;+aamuta+&#039;&lt;br&gt;&#039;;
                                }
                                var html=
                                    &#039;&lt;span style=&quot;font-weight: bold;&quot;&gt;Genomic DNA change:&lt;/span&gt;&lt;br&gt;&#039;+mutation.chromosome+&#039;:&#039;+mutation.chromosomeStart+&#039; &#039;+mutation.mutation+&#039;&lt;br&gt;&lt;br&gt;&#039;+
                                    &#039;&lt;span style=&quot;font-weight: bold;&quot;&gt;Mutation type:&lt;/span&gt;&lt;br&gt;&#039;+mutation.mutationType+&#039;&lt;br&gt;&lt;br&gt;&#039;+
                                    &#039;&lt;span style=&quot;font-weight: bold;&quot;&gt;Mutation Consequences:&lt;/span&gt;&lt;br&gt;&#039;+consequences+&#039;&lt;br&gt;&#039;;

                                $(div).qtip({
                                    content: {text: html, title: mutation.id},
                                    position: {target: &quot;mouse&quot;, adjust: {x: 25, y: 15}},
                                    style: { width: true, classes: &#039;qtip-bootstrap&#039;}
                                });
                                $(div).append(img);
                                return div;
                            };
                            for(var i in mutationsArray){
                                var div = mDiv(mutationsArray[i]);
                                $(&#039;#gv-adv-search-table&#039;).append(div);
                            }


                            if(advSearchMutationGenomeViewer != null) {
                                if(mutationsArray[0]) {
                                    var mut = mutationsArray[0];
                                    console.log(mut);
                                    regionObj = new Region({chromosome: mut.chromosome, start: parseInt(mut.chromosomeStart), end: parseInt(mut.chromosomeStart)});
                                    advSearchMutationGenomeViewer.setRegion(regionObj);
                                }
                                return;
                            }

                            window.projectFilterValues = [];
                            window.checkValues = [];
                            window.virtualTrackFilters = [];


                            advSearchMutationGenomeViewer = new GenomeViewer({
                                width: $(&quot;.col-span-9&quot;).outerWidth(),
                                height: 580,
                                region: regionObj,
                                defaultRegion: regionObj,
                                sidePanel: false,
                                targetId: &#039;gv-adv-search-mutation&#039;,
                                autoRender: true,
                                trackListTitle: &#039;&#039;,
                                drawNavigationBar: false,
                                drawKaryotypePanel: false,
                                drawChromosomePanel: false,
                                drawRegionOverviewPanel: false,
                                version: &#039;&#039;
                            });


                            var navigationBar = new IcgcNavigationBar({
                                handlers: {
                                    &#039;region:change&#039;: function (event) {
                                        Utils.setMinRegion(event.region, advSearchMutationGenomeViewer.getSVGCanvasWidth())
                                        advSearchMutationGenomeViewer.trigger(&#039;region:change&#039;, event);
                                    },
                                    &#039;size:small&#039;: function (event) {
                                        advSearchMutationGenomeViewer.setWidth(parseInt($(&#039;#gv-adv-search-mutation&#039;).parent().width()));
                                    },
                                    &#039;size:big&#039;: function (event) {
                                        advSearchMutationGenomeViewer.setWidth(parseInt($(window).width() - 30));
                                    },
//                                    &#039;fullscreen:click&#039;: function (event) {
//                                        if (advSearchMutationGenomeViewer.fullscreen) {
//                                            $(advSearchMutationGenomeViewer.div).css({width: &#039;auto&#039;});
//                                            Utils.cancelFullscreen();//no need to pass the dom object;
//                                            advSearchMutationGenomeViewer.fullscreen = false;
//                                        } else {
//                                            $(advSearchMutationGenomeViewer.div).css({width: screen.width});
//                                            Utils.launchFullScreen(advSearchMutationGenomeViewer.div);
//                                            advSearchMutationGenomeViewer.fullscreen = true;
//                                        }
//                                    },
                                    &#039;restoreDefaultRegion:click&#039;: function (event) {
                                        Utils.setMinRegion(advSearchMutationGenomeViewer.defaultRegion, advSearchMutationGenomeViewer.getSVGCanvasWidth());
                                        event.region = advSearchMutationGenomeViewer.defaultRegion;
                                        advSearchMutationGenomeViewer.trigger(&#039;region:change&#039;, event);
                                    }
                                }
                            });
                            advSearchMutationGenomeViewer.on(&#039;region:change&#039;, function (event) {
                                if (event.sender != navigationBar) {
                                    navigationBar.setRegion(event.region);
                                }
                            });
                            advSearchMutationGenomeViewer.on(&#039;region:move&#039;, function (event) {
                                if (event.sender != navigationBar) {
                                    navigationBar.moveRegion(event.region);
                                }
                            });
                            advSearchMutationGenomeViewer.on(&#039;width:change&#039;, function (event) {
                                navigationBar.setWidth(event.width);
                            });

                            advSearchMutationGenomeViewer.draw();

                            advSearchMutationGenomeViewer.setNavigationBar(navigationBar);
                            navigationBar.draw();

//                    // Making some panel not visible
//                    advSearchMutationGenomeViewer.navigationBar.setVisible({&#039;species&#039;:false, &#039;chromosomes&#039;: false, &#039;location&#039;: false, &#039;movement&#039;: false, &#039;search&#039;: false});
//                            advSearchMutationGenomeViewer.setKaryotypePanelVisible(false);
//                            advSearchMutationGenomeViewer.setChromosomePanelVisible(false);
//                            advSearchMutationGenomeViewer.setRegionOverviewPanelVisible(false);

                            // Adding tracks required
                            this.sequence = new SequenceTrack({
                                targetId: null,
                                id: 1,
                                title: &#039;Sequence&#039;,
                                histogramZoom: 20,
                                transcriptZoom: 50,
                                height: 30,
                                visibleRange: {start: 100, end: 100},
                                featureTypes: FEATURE_TYPES,

                                renderer: new SequenceRenderer(),

                                dataAdapter: new SequenceAdapter({
                                    category: &quot;genomic&quot;,
                                    subCategory: &quot;region&quot;,
                                    resource: &quot;sequence&quot;,
                                    species: advSearchMutationGenomeViewer.species,
                                    featureCache: {
                                        gzip: true,
                                        chunkSize: 1000
                                    }
                                })
                            });

                            advSearchMutationGenomeViewer.addTrack(this.sequence);

                            this.icgcGeneTrack = new IcgcGeneTrack({
                                targetId: null,
                                id: 2,
                                title: &#039;ICGC Genes&#039;,
                                histogramZoom: 20,
                                transcriptZoom: 50,
                                height: 260,
                                visibleRange: {start: 0, end: 100},
                                featureTypes: FEATURE_TYPES,

//                                renderer: new IcgcGeneRenderer(),
                                renderer: new GeneRenderer(),

                                dataAdapter: new IcgcGeneAdapter({
                                    category: &quot;genomic&quot;,
                                    subCategory: &quot;region&quot;,
                                    resource: &quot;snp&quot;,
                                    params: {
                                        exclude: &#039;transcriptVariations,xrefs,samples&#039;
                                    },
                                    species: advSearchMutationGenomeViewer.species,
                                    featureCache: {
                                        gzip: true,
                                        chunkSize: 10000
                                    },
                                    filters: {},
                                    options: {},
                                    featureConfig: FEATURE_CONFIG.gene
                                })
                            });
                            advSearchMutationGenomeViewer.addTrack(this.icgcGeneTrack);


                            this.icgcMutationsTrack = new IcgcMutationTrack({
                                targetId: null,
                                id: 3,
                                title: &#039;ICGC Mutations&#039;,
                                histogramZoom: 70,
                                labelZoom: 80,
                                height: 300,
                                visibleRange: {start: 0, end: 100},
//                        featureTypes:FEATURE_TYPES,

                                renderer: new FeatureRenderer({
                                    label: function (f) {
                                        return f.id;
                                    },
                                    tooltipTitle: function (f) {
                                        return &#039;ICGC mutation&#039; + &#039; - &lt;span class=&quot;ok&quot;&gt;&#039; + f.id + &#039;&lt;/span&gt;&#039;;
                                    },
                                    tooltipText: function (f) {
                                        var consequences = &#039;&#039;;
                                        for (var i = 0; i &lt; f.consequences.length; i++) {
                                            consequences += &#039;&amp;nbsp;&#039; + f.consequences[i] + &#039;&lt;br&gt;&#039;;
                                        }
                                        return &#039;mutation:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutation + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;reference allele:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.refGenAllele + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;mutation type:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.mutationType + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;project info:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + f.projectInfo[0] + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;consequences:&lt;br&gt;&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;&#039; + consequences + &#039;&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;source:&amp;nbsp;&lt;span class=&quot;ssel&quot;&gt;ICGC&lt;/span&gt;&lt;br&gt;&#039; +
                                            &#039;start-end:&amp;nbsp;&lt;span class=&quot;emph&quot;&gt;&#039; + f.start + &#039;-&#039; + f.end + &#039;&lt;/span&gt;&lt;br&gt;&#039;;
                                    },
                                    color: function (feat) {
                                        switch (feat.mutationType) {
                                            case &#039;single base substitution&#039;:
                                                return &#039;Chartreuse&#039;;
                                            case &#039;insertion of &lt;=200bp&#039;:
                                                return &#039;orange&#039;;
                                            case &#039;deletion of &lt;=200bp&#039;:
                                                return &#039;red&#039;;
                                            case &#039;multiple base substitution (&gt;=2bp and &lt;=200bp)&#039;:
                                                return &#039;lightblue&#039;;
                                            default:
                                                return &#039;black&#039;;
                                        }
                                    },
                                    infoWidgetId: &quot;mutationCds&quot;,
                                    height: 8,
                                    histogramColor: &quot;orange&quot;,
                                    handlers: {
                                        &#039;feature:mouseover&#039;: function (e) {
//                                    console.log(e)
                                        },
                                        &#039;feature:click&#039;: function (e) {
                                            console.log(e)
                                        }
                                    }
                                }),
//                                http://localhost:9001/api/mutations?filters={%22mutation%22:{%22mutation_type%22:[%22insertion%20of%20%3C=200bp%22,%22deletion%20of%20%3C=200bp%22]}}
                                dataAdapter: new IcgcMutationAdapter({
                                    category: &quot;genomic&quot;,
                                    subCategory: &quot;region&quot;,
                                    resource: &quot;snp&quot;,
                                    params: {
                                        filters: &#039;{&quot;mutation&quot;:{%22mutation_type%22:[%22insertion%20of%20%3C=200bp%22,%22deletion%20of%20%3C=200bp%22]}}&#039;
                                    },
                                    species: advSearchMutationGenomeViewer.species,
                                    featureCache: {
                                        gzip: true,
                                        chunkSize: 5000
                                    },
                                    filters: gvSearchMutationFilter,
                                    options: {},
                                    featureConfig: FEATURE_CONFIG.snp
                                })
                            });
                            advSearchMutationGenomeViewer.addTrack(this.icgcMutationsTrack);

                        });
                    }
                });
            }
        };
    });</textarea>
  </div>
</div>

<footer class="footer">
  <div class="container">
    <p>.</p>
  </div>
</footer>

<script type="text/html" id="complexity-popover-template">
  <div class="complexity-notice">
    Complexity : {{ complexity.cyclomatic }} <br>
    Length : {{ complexity.halstead.length }} <br>
    Difficulty : {{ complexity.halstead.difficulty.toFixed(2) }} <br>
    Est # bugs : {{ complexity.halstead.bugs.toFixed(2) }}<br>
  </div>
</script>

<script type="text/javascript" src="../../assets/scripts/bundles/core-bundle.js"></script>
<script type="text/javascript" src="../../assets/scripts/bundles/codemirror.js"></script>
<script type="text/javascript" src="../../assets/scripts/codemirror.markpopovertext.js"></script>
<script type="text/javascript" src="report.js"></script>
<script type="text/javascript" src="report.history.js"></script>
<script type="text/javascript" src="../../assets/scripts/plato-file.js"></script>
</body>
</html>
